{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "import os \n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "bucket = os.getenv(\"WORKSPACE_BUCKET\")\n",
    "\n",
    "USER_NAME = os.getenv('OWNER_EMAIL').split('@')[0].replace('.','-')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: USER_NAME=williamsjacr\n"
     ]
    }
   ],
   "source": [
    "%env USER_NAME={USER_NAME}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: dsub in /opt/conda/lib/python3.10/site-packages (0.4.13)\n",
      "Collecting dsub\n",
      "  Downloading dsub-0.5.0-py3-none-any.whl.metadata (32 kB)\n",
      "Requirement already satisfied: funcsigs==1.0.2 in /opt/conda/lib/python3.10/site-packages (from dsub) (1.0.2)\n",
      "Requirement already satisfied: google-api-core<=2.19.0,>=2.7.3 in /opt/conda/lib/python3.10/site-packages (from dsub) (2.11.1)\n",
      "Requirement already satisfied: google-api-python-client<=2.131.0,>=2.47.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (2.131.0)\n",
      "Requirement already satisfied: google-auth-httplib2<=0.2.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (0.2.0)\n",
      "Requirement already satisfied: google-auth<=2.29.0,>=2.6.6 in /opt/conda/lib/python3.10/site-packages (from dsub) (2.22.0)\n",
      "Requirement already satisfied: google-cloud-batch<=0.17.20 in /opt/conda/lib/python3.10/site-packages (from dsub) (0.17.20)\n",
      "Requirement already satisfied: httplib2<=0.22.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (0.21.0)\n",
      "Requirement already satisfied: mock<=5.1.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (5.1.0)\n",
      "Requirement already satisfied: parameterized<=0.9.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (0.9.0)\n",
      "Requirement already satisfied: protobuf<=5.26.0,>=3.19.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (3.19.6)\n",
      "Requirement already satisfied: pyasn1-modules<=0.4.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (0.2.7)\n",
      "Requirement already satisfied: pyasn1<=0.6.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (0.4.8)\n",
      "Requirement already satisfied: python-dateutil<=2.9.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (2.8.2)\n",
      "Requirement already satisfied: pytz<=2024.1 in /opt/conda/lib/python3.10/site-packages (from dsub) (2023.3)\n",
      "Requirement already satisfied: pyyaml<=6.0.1 in /opt/conda/lib/python3.10/site-packages (from dsub) (6.0.1)\n",
      "Requirement already satisfied: rsa<=4.9 in /opt/conda/lib/python3.10/site-packages (from dsub) (4.9)\n",
      "Requirement already satisfied: tabulate<=0.9.0 in /opt/conda/lib/python3.10/site-packages (from dsub) (0.9.0)\n",
      "Requirement already satisfied: tenacity<=8.2.3 in /opt/conda/lib/python3.10/site-packages (from dsub) (8.2.3)\n",
      "Requirement already satisfied: uritemplate<=4.1.1 in /opt/conda/lib/python3.10/site-packages (from dsub) (3.0.1)\n",
      "Requirement already satisfied: googleapis-common-protos<2.0.dev0,>=1.56.2 in /opt/conda/lib/python3.10/site-packages (from google-api-core<=2.19.0,>=2.7.3->dsub) (1.60.0)\n",
      "Requirement already satisfied: requests<3.0.0.dev0,>=2.18.0 in /opt/conda/lib/python3.10/site-packages (from google-api-core<=2.19.0,>=2.7.3->dsub) (2.32.3)\n",
      "Requirement already satisfied: cachetools<6.0,>=2.0.0 in /opt/conda/lib/python3.10/site-packages (from google-auth<=2.29.0,>=2.6.6->dsub) (4.2.4)\n",
      "Requirement already satisfied: six>=1.9.0 in /opt/conda/lib/python3.10/site-packages (from google-auth<=2.29.0,>=2.6.6->dsub) (1.16.0)\n",
      "Requirement already satisfied: urllib3<2.0 in /opt/conda/lib/python3.10/site-packages (from google-auth<=2.29.0,>=2.6.6->dsub) (1.26.15)\n",
      "Requirement already satisfied: proto-plus<2.0.0dev,>=1.22.3 in /opt/conda/lib/python3.10/site-packages (from google-cloud-batch<=0.17.20->dsub) (1.22.3)\n",
      "Requirement already satisfied: pyparsing!=3.0.0,!=3.0.1,!=3.0.2,!=3.0.3,<4,>=2.4.2 in /opt/conda/lib/python3.10/site-packages (from httplib2<=0.22.0->dsub) (3.0.9)\n",
      "Requirement already satisfied: grpcio<2.0dev,>=1.33.2 in /opt/conda/lib/python3.10/site-packages (from google-api-core[grpc]!=2.0.*,!=2.1.*,!=2.10.*,!=2.2.*,!=2.3.*,!=2.4.*,!=2.5.*,!=2.6.*,!=2.7.*,!=2.8.*,!=2.9.*,<3.0.0dev,>=1.34.1->google-cloud-batch<=0.17.20->dsub) (1.48.1)\n",
      "Requirement already satisfied: grpcio-status<2.0.dev0,>=1.33.2 in /opt/conda/lib/python3.10/site-packages (from google-api-core[grpc]!=2.0.*,!=2.1.*,!=2.10.*,!=2.2.*,!=2.3.*,!=2.4.*,!=2.5.*,!=2.6.*,!=2.7.*,!=2.8.*,!=2.9.*,<3.0.0dev,>=1.34.1->google-cloud-batch<=0.17.20->dsub) (1.48.1)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /opt/conda/lib/python3.10/site-packages (from requests<3.0.0.dev0,>=2.18.0->google-api-core<=2.19.0,>=2.7.3->dsub) (3.2.0)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /opt/conda/lib/python3.10/site-packages (from requests<3.0.0.dev0,>=2.18.0->google-api-core<=2.19.0,>=2.7.3->dsub) (3.4)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/conda/lib/python3.10/site-packages (from requests<3.0.0.dev0,>=2.18.0->google-api-core<=2.19.0,>=2.7.3->dsub) (2024.2.2)\n",
      "Downloading dsub-0.5.0-py3-none-any.whl (185 kB)\n",
      "Installing collected packages: dsub\n",
      "Successfully installed dsub-0.5.0\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m24.2\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m24.3.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip3 install --upgrade dsub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing /home/jupyter/aou_dsub.bash\n"
     ]
    }
   ],
   "source": [
    "%%writefile ~/aou_dsub.bash\n",
    "#!/bin/bash\n",
    "function aou_dsub () {\n",
    "\n",
    "  # Get a shorter username to leave more characters for the job name.\n",
    "  local DSUB_USER_NAME=\"$(echo \"${OWNER_EMAIL}\" | cut -d@ -f1)\"\n",
    "\n",
    "  # For AoU RWB projects network name is \"network\".\n",
    "  local AOU_NETWORK=network\n",
    "  local AOU_SUBNETWORK=subnetwork\n",
    "\n",
    "  dsub \\\n",
    "      --provider google-cls-v2 \\\n",
    "      --user-project \"${GOOGLE_PROJECT}\"\\\n",
    "      --project \"${GOOGLE_PROJECT}\"\\\n",
    "      --image 'marketplace.gcr.io/google/ubuntu1804:latest' \\\n",
    "      --network \"${AOU_NETWORK}\" \\\n",
    "      --subnetwork \"${AOU_SUBNETWORK}\" \\\n",
    "      --service-account \"$(gcloud config get-value account)\" \\\n",
    "      --user \"${DSUB_USER_NAME}\" \\\n",
    "      --regions us-central1 \\\n",
    "      --logging \"${WORKSPACE_BUCKET}/dsub/logs/{job-name}/{user-id}/$(date +'%Y%m%d/%H%M%S')/{job-id}-{task-id}-{task-attempt}.log\" \\\n",
    "      \"$@\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "!echo source ~/aou_dsub.bash >> ~/.bashrc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting NullModel.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile NullModel.R\n",
    "rm(list = ls())\n",
    "\n",
    "all_phenotype_File <- commandArgs(TRUE)[1]\n",
    "print(all_phenotype_File)\n",
    "\n",
    "gds_file <- commandArgs(TRUE)[2]\n",
    "print(gds_file)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[3]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "library(gdsfmt)\n",
    "library(SeqArray)\n",
    "library(SeqVarTools)\n",
    "library(dplyr)\n",
    "library(STAAR)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "library(Matrix)\n",
    "library(SCANG)\n",
    "library(STAARpipeline)\n",
    "\n",
    "trait <- \"TC\"\n",
    "\n",
    "agds.path <- gds_file\n",
    "\n",
    "genofile <- seqOpen(agds.path)\n",
    "\n",
    "id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "\n",
    "seqClose(genofile)\n",
    "\n",
    "print(str(id.genotype))\n",
    "\n",
    "all_phenotypes <- read.csv(all_phenotype_File)\n",
    "\n",
    "colnames(all_phenotypes)[10] <- \"sex\"\n",
    "\n",
    "print(str(all_phenotypes))\n",
    "\n",
    "all_phenotypes <- all_phenotypes[all_phenotypes$IID %in% id.genotype,]\n",
    "\n",
    "print(nrow(all_phenotypes))\n",
    "\n",
    "all_phenotypes <- all_phenotypes[!is.na(all_phenotypes[,trait]),]\n",
    "  \n",
    "print(nrow(all_phenotypes))\n",
    "\n",
    "all_phenotypes$age2 <- all_phenotypes$age^2\n",
    "\n",
    "all_phenotypes$TCadj_resid <- resid(lm(TC~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10, data = all_phenotypes))\n",
    "\n",
    "all_phenotypes$TCadj_norm <- sd(all_phenotypes$TC)*scale(qnorm((rank(all_phenotypes$TCadj_resid,na.last=\"keep\")-0.5)/length(all_phenotypes$TCadj_resid)))\n",
    "\n",
    "obj.STAAR.UKB <- fit_nullmodel(TCadj_norm~age+age2+sex+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10, data = all_phenotypes,id = \"IID\",kins = NULL,family = gaussian(link = \"identity\"))\n",
    "  \n",
    "save(obj.STAAR.UKB,file = paste0(OUTPUT_PATH,\"/TC_Null_Model.RData\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "  '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/MetaSTAAR/NullModel.R\",\n",
    "  '--input all_phenotype_File'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/JW/AoU_Phenotypes/Results/all_phenotypes.csv\",\n",
    "  '--input gds_file'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataAGDS/exome_v7.1/exome.chr21.gds\",\n",
    "  '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/MetaSTAAR/\",\n",
    "  check.names = FALSE\n",
    "))\n",
    "\n",
    "write.table(tasks,\n",
    "            file=\"score_task.txt\",\n",
    "            row.names=F, col.names=T,\n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting NullModel.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile NullModel.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${all_phenotype_File} ${gds_file} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://NullModel.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][  1.4 KiB/  1.4 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/1.4 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp NullModel.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/MetaSTAAR/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: nullmodel--williamsjacr--240910-152518-10\n",
      "  job-name: nullmodel\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9054957783243937729\n",
      "Launched job-id: nullmodel--williamsjacr--240910-152518-10\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'nullmodel--williamsjacr--240910-152518-10' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'nullmodel--williamsjacr--240910-152518-10' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --timeout \"48h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script NullModel.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name      Task  Status    Last Update\r\n",
      "----------  ------  --------  --------------\r\n",
      "nullmodel        1  Success   09-10 15:30:18\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'nullmodel--williamsjacr--240910-152518-10' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "code_folding": [
     36,
     39,
     110,
     113,
     186
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Meta_Coding_AoU.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile Meta_Coding_AoU.R\n",
    "rm(list=ls())\n",
    "gc()\n",
    "\n",
    "arrayid <- as.numeric(commandArgs(TRUE)[1])\n",
    "\n",
    "Null_Model <- commandArgs(TRUE)[2]\n",
    "print(Null_Model)\n",
    "\n",
    "Annotation_name_catalog <- commandArgs(TRUE)[3]\n",
    "print(Annotation_name_catalog)\n",
    "\n",
    "GDS_File <- commandArgs(TRUE)[4]\n",
    "print(GDS_File)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[5]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "time <- system.time({\n",
    "  ## load required packages\n",
    "  library(gdsfmt)\n",
    "  library(SeqArray)\n",
    "  library(SeqVarTools)\n",
    "  library(STAAR)\n",
    "  library(STAARpipeline)\n",
    "  library(readr)\n",
    "  library(dplyr)\n",
    "  library(stringr)\n",
    "  \n",
    "  library(devtools)\n",
    "#  build(R_PACKAGE,path = \"\")\n",
    "#  install.packages(\"MetaSTAAR_0.9.6.3.tar.gz\", repos = NULL, type=\"source\")\n",
    "  library(MetaSTAAR)\n",
    "  library(expm)\n",
    "  library(Matrix)\n",
    "  \n",
    "  MetaSTAARlite_worker_sumstat <- function(genotype,obj_nullmodel,variant_info,qc_label=NULL,\n",
    "                                           annotation_phred=NULL,for_individual_analysis=FALSE){\n",
    "    \n",
    "    if(!inherits(genotype, \"matrix\") && !inherits(genotype, \"Matrix\")){\n",
    "      stop(\"genotype is not a matrix!\")\n",
    "    }\n",
    "    \n",
    "    if(inherits(genotype, \"sparseMatrix\")){\n",
    "      genotype <- as.matrix(genotype)\n",
    "    }\n",
    "    \n",
    "    if(dim(genotype)[2] != dim(variant_info)[1]){\n",
    "      stop(paste0(\"Dimensions don't match for genotype and variant_info!\"))\n",
    "    }\n",
    "    \n",
    "    if(!is.null(qc_label) && dim(variant_info)[1] != length(qc_label)){\n",
    "      stop(paste0(\"Dimensions don't match for variant_info and qc_label!\"))\n",
    "    }\n",
    "    \n",
    "    if(!is.null(annotation_phred) && dim(annotation_phred)[1] != dim(variant_info)[1]){\n",
    "      stop(paste0(\"Dimensions don't match for annotation_phred and variant_info!\"))\n",
    "    }\n",
    "    \n",
    "    N <- dim(genotype)[1]\n",
    "    alt_AC <- as.integer(colSums(2 - genotype))\n",
    "    MAC <- as.integer(pmin(alt_AC, 2 * N - alt_AC))\n",
    "    genotype <- matrix_flip(genotype)\n",
    "    MAF <- genotype$MAF\n",
    "    variant_label <- as.vector(MAF>0)\n",
    "    Geno <- genotype$Geno[,variant_label,drop=FALSE]\n",
    "    Geno <- as(Geno,\"dgCMatrix\")\n",
    "    rm(genotype)\n",
    "    gc()\n",
    "    \n",
    "    if(obj_nullmodel$relatedness){\n",
    "      if(!obj_nullmodel$sparse_kins){\n",
    "        stop(paste0(\"Please use a sparse genetic relatedness matrix when fitting the null model!\"))\n",
    "      }\n",
    "    }else{\n",
    "      if(obj_nullmodel$family[1] == \"binomial\"){\n",
    "        obj_nullmodel$Sigma_i <- Diagonal(x = obj_nullmodel$weights)\n",
    "      }else if(obj_nullmodel$family[1] == \"gaussian\"){\n",
    "        obj_nullmodel$Sigma_i <- Diagonal(length(obj_nullmodel$y)) / summary(obj_nullmodel)$dispersion\n",
    "      }\n",
    "      obj_nullmodel$Sigma_iX <- obj_nullmodel$Sigma_i %*% model.matrix(obj_nullmodel)\n",
    "      obj_nullmodel$cov <- solve(t(model.matrix(obj_nullmodel)) %*% obj_nullmodel$Sigma_i %*% model.matrix(obj_nullmodel))\n",
    "      obj_nullmodel$scaled.residuals <- (obj_nullmodel$y - obj_nullmodel$fitted.values) / summary(obj_nullmodel)$dispersion\n",
    "    }\n",
    "    \n",
    "    GTSinvX_cov <- as(t(Geno) %*% obj_nullmodel$Sigma_iX,\"matrix\") %*% sqrtm(obj_nullmodel$cov)\n",
    "    #V <- diag(t(Geno) %*% obj_nullmodel$Sigma_i %*% Geno - GTSinvX_cov %*% t(GTSinvX_cov))\n",
    "    V <- colSums(Geno * (obj_nullmodel$Sigma_i %*% Geno)) - rowSums(GTSinvX_cov^2) # faster for large number of variants\n",
    "    U <- as.vector(t(Geno) %*% obj_nullmodel$scaled.residuals)\n",
    "    rm(Geno)\n",
    "    gc()\n",
    "    \n",
    "    if (!is.null(qc_label)){\n",
    "      sumstat <- data.frame(variant_info[,c(\"chr\",\"pos\",\"ref\",\"alt\")],qc_label,alt_AC,MAC,MAF,N)\n",
    "    }else{\n",
    "      sumstat <- data.frame(variant_info[,c(\"chr\",\"pos\",\"ref\",\"alt\")],alt_AC,MAC,MAF,N)\n",
    "    }\n",
    "    if (!is.null(annotation_phred)){\n",
    "      sumstat <- cbind(sumstat,annotation_phred)\n",
    "    }\n",
    "    sumstat <- sumstat[variant_label,]\n",
    "    if(!for_individual_analysis){\n",
    "      sumstat <- cbind(sumstat,U,V,GTSinvX_cov)\n",
    "    }else{\n",
    "      sumstat <- cbind(sumstat,U,V)\n",
    "    }\n",
    "    \n",
    "    return(sumstat)\n",
    "  }\n",
    "  \n",
    "  MetaSTAARlite_worker_cov <- function(genotype,obj_nullmodel,cov_maf_cutoff=0.05,\n",
    "                                       qc_label=NULL,signif.digits=3){\n",
    "    \n",
    "    if(!inherits(genotype, \"matrix\") && !inherits(genotype, \"Matrix\")){\n",
    "      stop(\"genotype is not a matrix!\")\n",
    "    }\n",
    "    \n",
    "    if(cov_maf_cutoff < 0 | cov_maf_cutoff > 0.5){\n",
    "      stop(\"cov_maf_cutoff should be a number between 0 and 0.5!\")\n",
    "    }\n",
    "    \n",
    "    if(cov_maf_cutoff == 0.5){\n",
    "      cov_maf_cutoff <- 0.5 + 1e-16\n",
    "    }\n",
    "    \n",
    "    if(inherits(genotype, \"sparseMatrix\")){\n",
    "      MAF <- colMeans(genotype)/2\n",
    "      if (!is.null(qc_label)){\n",
    "        RV_label <- as.vector((MAF<cov_maf_cutoff)&(MAF>0)&(qc_label==\"PASS\"))\n",
    "      }else{\n",
    "        RV_label <- as.vector((MAF<cov_maf_cutoff)&(MAF>0))\n",
    "      }\n",
    "      Geno_rare <- genotype[,RV_label,drop=FALSE]\n",
    "      Geno_rare <- as(Geno_rare,\"dgCMatrix\")\n",
    "    }else{\n",
    "      genotype <- matrix_flip(genotype)\n",
    "      MAF <- genotype$MAF\n",
    "      if (!is.null(qc_label)){\n",
    "        RV_label <- as.vector((MAF<cov_maf_cutoff)&(MAF>0)&(qc_label==\"PASS\"))\n",
    "      }else{\n",
    "        RV_label <- as.vector((MAF<cov_maf_cutoff)&(MAF>0))\n",
    "      }\n",
    "      Geno_rare <- genotype$Geno[,RV_label,drop=FALSE]\n",
    "      Geno_rare <- as(Geno_rare,\"dgCMatrix\")\n",
    "    }\n",
    "    \n",
    "    rm(genotype,MAF)\n",
    "    gc()\n",
    "    \n",
    "    if(obj_nullmodel$relatedness){\n",
    "      if(!obj_nullmodel$sparse_kins){\n",
    "        stop(paste0(\"Please use a sparse genetic relatedness matrix when fitting the null model!\"))\n",
    "      }\n",
    "    }else{\n",
    "      if(obj_nullmodel$family[1] == \"binomial\"){\n",
    "        obj_nullmodel$Sigma_i <- Diagonal(x = obj_nullmodel$weights)\n",
    "      }else if(obj_nullmodel$family[1] == \"gaussian\"){\n",
    "        obj_nullmodel$Sigma_i <- Diagonal(length(obj_nullmodel$y)) / summary(obj_nullmodel)$dispersion\n",
    "      }\n",
    "      obj_nullmodel$Sigma_iX <- obj_nullmodel$Sigma_i %*% model.matrix(obj_nullmodel)\n",
    "      obj_nullmodel$cov <- solve(t(model.matrix(obj_nullmodel)) %*% obj_nullmodel$Sigma_i %*% model.matrix(obj_nullmodel))\n",
    "      obj_nullmodel$scaled.residuals <- (obj_nullmodel$y - obj_nullmodel$fitted.values) / summary(obj_nullmodel)$dispersion\n",
    "    }\n",
    "    \n",
    "    GTSinvG_rare <- t(obj_nullmodel$Sigma_i %*% Geno_rare) %*% Geno_rare\n",
    "    rm(Geno_rare)\n",
    "    gc()\n",
    "    \n",
    "    GTSinvG_rare <- as(GTSinvG_rare,\"TsparseMatrix\")\n",
    "    remove_ind <- (GTSinvG_rare@j < GTSinvG_rare@i) # Be careful about multi-allelic issue\n",
    "    GTSinvG_rare@i <- GTSinvG_rare@i[!remove_ind]\n",
    "    GTSinvG_rare@j <- GTSinvG_rare@j[!remove_ind]\n",
    "    GTSinvG_rare@x <- GTSinvG_rare@x[!remove_ind]\n",
    "    rm(remove_ind)\n",
    "    gc()\n",
    "    GTSinvG_rare <- as(GTSinvG_rare,\"CsparseMatrix\")\n",
    "    \n",
    "    ### Create a version of GTSinvG_rare with rounded significant digits\n",
    "    if(!is.null(signif.digits)){\n",
    "      GTSinvG_rare <- signif(GTSinvG_rare, digits = signif.digits)\n",
    "    }\n",
    "    \n",
    "    return(GTSinvG_rare)\n",
    "  }\n",
    "  \n",
    "  \n",
    "  coding_MetaSTAARlite_worker <- function(chr,gene_name,genofile,obj_nullmodel,genes,known_loci=NULL,\n",
    "                                          cov_maf_cutoff=0.05,signif.digits=NULL,\n",
    "                                          QC_label=\"annotation/filter\",check_qc_label=FALSE,variant_type=c(\"SNV\",\"Indel\",\"variant\"),\n",
    "                                          Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                                          Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                                          silent=FALSE){\n",
    "    \n",
    "    ## evaluate choices\n",
    "    variant_type <- match.arg(variant_type)\n",
    "    \n",
    "    phenotype.id <- obj_nullmodel$id_include\n",
    "    \n",
    "    filter <- seqGetData(genofile, QC_label)\n",
    "    if(variant_type==\"variant\")\n",
    "    {\n",
    "      if(check_qc_label)\n",
    "      {\n",
    "        SNVlist <- TRUE\n",
    "      }else\n",
    "      {\n",
    "        SNVlist <- filter == \"PASS\"\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    if(variant_type==\"SNV\")\n",
    "    {\n",
    "      if(check_qc_label)\n",
    "      {\n",
    "        SNVlist <- isSNV(genofile)\n",
    "      }else\n",
    "      {\n",
    "        SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    if(variant_type==\"Indel\")\n",
    "    {\n",
    "      if(check_qc_label)\n",
    "      {\n",
    "        SNVlist <- !isSNV(genofile)\n",
    "      }else\n",
    "      {\n",
    "        SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    position <- as.numeric(seqGetData(genofile, \"position\"))\n",
    "    variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "    \n",
    "    if(!is.null(known_loci))\n",
    "    {\n",
    "      allele <- seqGetData(genofile, \"allele\")\n",
    "      \n",
    "      loc_SNV <- c()\n",
    "      for (i in 1:dim(known_loci)[1]){\n",
    "        loc_SNV <- c(loc_SNV,which((position==known_loci$POS[i])&(allele==paste0(known_loci$REF[i],\",\",known_loci$ALT[i]))))\n",
    "      }\n",
    "      \n",
    "      seqSetFilter(genofile,variant.id=variant.id[loc_SNV],sample.id=phenotype.id)\n",
    "      \n",
    "      G_SNV <- seqGetData(genofile, \"$dosage\")\n",
    "      \n",
    "      if (!is.null(G_SNV)){\n",
    "        id.SNV <- seqGetData(genofile,\"sample.id\")\n",
    "        id.SNV.match <- rep(0,length(id.SNV))\n",
    "        \n",
    "        for(i in 1:length(id.SNV))\n",
    "        {\n",
    "          id.SNV.match[i] <- which.max(id.SNV==phenotype.id[i])\n",
    "        }\n",
    "        \n",
    "        G_SNV <- G_SNV[id.SNV.match,,drop=FALSE]\n",
    "        G_SNV_MAF <- apply(G_SNV, 2, function(x){sum(x[!is.na(x)])/2/sum(!is.na(x))})\n",
    "        for (i in 1:length(G_SNV_MAF)){\n",
    "          if (G_SNV_MAF[i] <= 0.5){\n",
    "            G_SNV[is.na(G_SNV[,i]),i] <- 0\n",
    "          }\n",
    "          else{\n",
    "            G_SNV[is.na(G_SNV[,i]),i] <- 2\n",
    "          }\n",
    "        }\n",
    "        \n",
    "        pos_adj <- as.integer(seqGetData(genofile, \"position\"))\n",
    "        ref_adj <- as.character(seqGetData(genofile, \"$ref\"))\n",
    "        alt_adj <- as.character(seqGetData(genofile, \"$alt\"))\n",
    "        variant_adj_info <- data.frame(chr,pos_adj,ref_adj,alt_adj)\n",
    "        colnames(variant_adj_info) <- c(\"chr\",\"pos\",\"ref\",\"alt\")\n",
    "        variant_adj_info\n",
    "        \n",
    "        seqResetFilter(genofile)\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    rm(filter)\n",
    "    gc()\n",
    "    \n",
    "    ### Gene\n",
    "    kk <- which(genes[,1]==gene_name)\n",
    "    \n",
    "    sub_start_loc <- genes[kk,3]\n",
    "    sub_end_loc <- genes[kk,4]\n",
    "    \n",
    "    is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "    variant.id.gene <- variant.id[is.in]\n",
    "    \n",
    "    rm(position)\n",
    "    gc()\n",
    "    \n",
    "    seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "    \n",
    "    ## Gencode_Exonic\n",
    "    GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "    ## Gencode\n",
    "    GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "    ## Meta.SVM.Pred\n",
    "    MetaSVM_pred <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"MetaSVM\")]))\n",
    "    \n",
    "    ################################################\n",
    "    #           Coding\n",
    "    ################################################\n",
    "    variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "    lof.in.coding <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")|(GENCODE.EXONIC.Category==\"nonsynonymous SNV\")|(GENCODE.EXONIC.Category==\"synonymous SNV\")\n",
    "    variant.id.gene <- variant.id.gene[lof.in.coding]\n",
    "    \n",
    "    seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "    \n",
    "    ## Gencode_Exonic\n",
    "    GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "    ## Gencode\n",
    "    GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "    ## Meta.SVM.Pred\n",
    "    MetaSVM_pred <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"MetaSVM\")]))\n",
    "    \n",
    "    ## Annotation\n",
    "    Anno.Int.PHRED.sub <- NULL\n",
    "    Anno.Int.PHRED.sub.name <- NULL\n",
    "    \n",
    "    if(variant_type==\"SNV\")\n",
    "    {\n",
    "      if(Use_annotation_weights)\n",
    "      {\n",
    "        for(k in 1:length(Annotation_name))\n",
    "        {\n",
    "          if(Annotation_name[k]%in%Annotation_name_catalog$name)\n",
    "          {\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "            Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "            \n",
    "            if(Annotation_name[k]==\"CADD\")\n",
    "            {\n",
    "              Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "            }\n",
    "            Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "          }\n",
    "        }\n",
    "        \n",
    "        Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "        colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    ################################################\n",
    "    #                  plof_ds\n",
    "    ################################################\n",
    "    variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")|((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "    variant.id.gene.category <- variant.id.gene[lof.in.plof]\n",
    "    \n",
    "    seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "    \n",
    "    pos <- as.integer(seqGetData(genofile, \"position\"))\n",
    "    ref <- as.character(seqGetData(genofile, \"$ref\"))\n",
    "    alt <- as.character(seqGetData(genofile, \"$alt\"))\n",
    "    if(check_qc_label){\n",
    "      qc_label <- as.character(seqGetData(genofile, QC_label))\n",
    "    }else{\n",
    "      qc_label <- NULL\n",
    "    }\n",
    "    \n",
    "    ## genotype id\n",
    "    id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "    # id.genotype.match <- rep(0,length(id.genotype))\n",
    "    \n",
    "    id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "    phenotype.id.merge <- data.frame(phenotype.id)\n",
    "    phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "    id.genotype.match <- phenotype.id.merge$index\n",
    "    \n",
    "    ## Genotype\n",
    "    Geno <- seqGetData(genofile, \"$dosage\")\n",
    "    Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "    \n",
    "    summary_stat_list <- list()\n",
    "    GTSinvG_rare_list <- list()\n",
    "    cov_cond_list <- list()\n",
    "    \n",
    "    summary_stat <- NULL\n",
    "    GTSinvG_rare <- NULL\n",
    "    cov_cond <- NULL\n",
    "    \n",
    "    if(!is.null(Geno))\n",
    "    {\n",
    "      # Summary statistics\n",
    "      genotype <- matrix_impute(Geno)\n",
    "      variant_info <- data.frame(chr,pos,ref,alt,row.names=NULL)\n",
    "      \n",
    "      ## Annotation\n",
    "      Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.plof,]\n",
    "      \n",
    "      try(summary_stat <- MetaSTAARlite_worker_sumstat(genotype,obj_nullmodel,variant_info,qc_label,\n",
    "                                                       Anno.Int.PHRED.sub.category),silent=silent)\n",
    "      \n",
    "      # Covariance matrices\n",
    "      genotype <- matrix_flip(Geno)$Geno\n",
    "      genotype <- as(genotype,\"dgCMatrix\")\n",
    "      \n",
    "      rm(Geno)\n",
    "      gc()\n",
    "      \n",
    "      try(GTSinvG_rare <- MetaSTAARlite_worker_cov(genotype,obj_nullmodel,cov_maf_cutoff,\n",
    "                                                   qc_label,signif.digits),silent=silent)\n",
    "      \n",
    "      # Covariance matrices for conditional analysis\n",
    "      if(!is.null(known_loci))\n",
    "      {\n",
    "        try(cov_cond <- MetaSTAARlite_worker_cov_cond(genotype,G_SNV,obj_nullmodel,variant_info,variant_adj_info),silent=silent)\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    summary_stat_list[[\"plof_ds\"]] <- summary_stat\n",
    "    GTSinvG_rare_list[[\"plof_ds\"]] <- GTSinvG_rare\n",
    "    cov_cond_list[[\"plof_ds\"]] <- cov_cond\n",
    "    \n",
    "    #####################################################\n",
    "    #                      plof\n",
    "    #####################################################\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")\n",
    "    variant.id.gene.category <- variant.id.gene[lof.in.plof]\n",
    "    \n",
    "    seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "    \n",
    "    pos <- as.integer(seqGetData(genofile, \"position\"))\n",
    "    ref <- as.character(seqGetData(genofile, \"$ref\"))\n",
    "    alt <- as.character(seqGetData(genofile, \"$alt\"))\n",
    "    if(check_qc_label){\n",
    "      qc_label <- as.character(seqGetData(genofile, QC_label))\n",
    "    }else{\n",
    "      qc_label <- NULL\n",
    "    }\n",
    "    \n",
    "    ## genotype id\n",
    "    id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "    # id.genotype.match <- rep(0,length(id.genotype))\n",
    "    \n",
    "    id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "    phenotype.id.merge <- data.frame(phenotype.id)\n",
    "    phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "    id.genotype.match <- phenotype.id.merge$index\n",
    "    \n",
    "    ## Genotype\n",
    "    Geno <- seqGetData(genofile, \"$dosage\")\n",
    "    Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "    \n",
    "    summary_stat <- NULL\n",
    "    GTSinvG_rare <- NULL\n",
    "    cov_cond <- NULL\n",
    "    \n",
    "    if(!is.null(Geno))\n",
    "    {\n",
    "      # Summary statistics\n",
    "      genotype <- matrix_impute(Geno)\n",
    "      variant_info <- data.frame(chr,pos,ref,alt,row.names=NULL)\n",
    "      \n",
    "      ## Annotation\n",
    "      Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.plof,]\n",
    "      \n",
    "      try(summary_stat <- MetaSTAARlite_worker_sumstat(genotype,obj_nullmodel,variant_info,qc_label,\n",
    "                                                       Anno.Int.PHRED.sub.category),silent=silent)\n",
    "      \n",
    "      # Covariance matrices\n",
    "      genotype <- matrix_flip(Geno)$Geno\n",
    "      genotype <- as(genotype,\"dgCMatrix\")\n",
    "      \n",
    "      rm(Geno)\n",
    "      gc()\n",
    "      \n",
    "      try(GTSinvG_rare <- MetaSTAARlite_worker_cov(genotype,obj_nullmodel,cov_maf_cutoff,\n",
    "                                                   qc_label,signif.digits),silent=silent)\n",
    "      \n",
    "      # Covariance matrices for conditional analysis\n",
    "      if(!is.null(known_loci))\n",
    "      {\n",
    "        try(cov_cond <- MetaSTAARlite_worker_cov_cond(genotype,G_SNV,obj_nullmodel,variant_info,variant_adj_info),silent=silent)\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    summary_stat_list[[\"plof\"]] <- summary_stat\n",
    "    GTSinvG_rare_list[[\"plof\"]] <- GTSinvG_rare\n",
    "    cov_cond_list[[\"plof\"]] <- cov_cond\n",
    "    \n",
    "    #############################################\n",
    "    #             synonymous\n",
    "    #############################################\n",
    "    lof.in.synonymous <- (GENCODE.EXONIC.Category==\"synonymous SNV\")\n",
    "    variant.id.gene.category <- variant.id.gene[lof.in.synonymous]\n",
    "    \n",
    "    seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "    \n",
    "    pos <- as.integer(seqGetData(genofile, \"position\"))\n",
    "    ref <- as.character(seqGetData(genofile, \"$ref\"))\n",
    "    alt <- as.character(seqGetData(genofile, \"$alt\"))\n",
    "    if(check_qc_label){\n",
    "      qc_label <- as.character(seqGetData(genofile, QC_label))\n",
    "    }else{\n",
    "      qc_label <- NULL\n",
    "    }\n",
    "    \n",
    "    ## genotype id\n",
    "    id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "    # id.genotype.match <- rep(0,length(id.genotype))\n",
    "    \n",
    "    id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "    phenotype.id.merge <- data.frame(phenotype.id)\n",
    "    phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "    id.genotype.match <- phenotype.id.merge$index\n",
    "    \n",
    "    ## Genotype\n",
    "    Geno <- seqGetData(genofile, \"$dosage\")\n",
    "    Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "    \n",
    "    summary_stat <- NULL\n",
    "    GTSinvG_rare <- NULL\n",
    "    cov_cond <- NULL\n",
    "    \n",
    "    if(!is.null(Geno))\n",
    "    {\n",
    "      # Summary statistics\n",
    "      genotype <- matrix_impute(Geno)\n",
    "      variant_info <- data.frame(chr,pos,ref,alt,row.names=NULL)\n",
    "      \n",
    "      ## Annotation\n",
    "      Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.synonymous,]\n",
    "      \n",
    "      try(summary_stat <- MetaSTAARlite_worker_sumstat(genotype,obj_nullmodel,variant_info,qc_label,\n",
    "                                                       Anno.Int.PHRED.sub.category),silent=silent)\n",
    "      \n",
    "      # Covariance matrices\n",
    "      genotype <- matrix_flip(Geno)$Geno\n",
    "      genotype <- as(genotype,\"dgCMatrix\")\n",
    "      \n",
    "      rm(Geno)\n",
    "      gc()\n",
    "      \n",
    "      try(GTSinvG_rare <- MetaSTAARlite_worker_cov(genotype,obj_nullmodel,cov_maf_cutoff,\n",
    "                                                   qc_label,signif.digits),silent=silent)\n",
    "      \n",
    "      # Covariance matrices for conditional analysis\n",
    "      if(!is.null(known_loci))\n",
    "      {\n",
    "        try(cov_cond <- MetaSTAARlite_worker_cov_cond(genotype,G_SNV,obj_nullmodel,variant_info,variant_adj_info),silent=silent)\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    summary_stat_list[[\"synonymous\"]] <- summary_stat\n",
    "    GTSinvG_rare_list[[\"synonymous\"]] <- GTSinvG_rare\n",
    "    cov_cond_list[[\"synonymous\"]] <- cov_cond\n",
    "    \n",
    "    #################################################\n",
    "    #        missense\n",
    "    #################################################\n",
    "    lof.in.missense <- (GENCODE.EXONIC.Category==\"nonsynonymous SNV\")\n",
    "    variant.id.gene.category <- variant.id.gene[lof.in.missense]\n",
    "    \n",
    "    seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "    \n",
    "    pos <- as.integer(seqGetData(genofile, \"position\"))\n",
    "    ref <- as.character(seqGetData(genofile, \"$ref\"))\n",
    "    alt <- as.character(seqGetData(genofile, \"$alt\"))\n",
    "    if(check_qc_label){\n",
    "      qc_label <- as.character(seqGetData(genofile, QC_label))\n",
    "    }else{\n",
    "      qc_label <- NULL\n",
    "    }\n",
    "    \n",
    "    ## genotype id\n",
    "    id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "    # id.genotype.match <- rep(0,length(id.genotype))\n",
    "    \n",
    "    id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "    phenotype.id.merge <- data.frame(phenotype.id)\n",
    "    phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "    id.genotype.match <- phenotype.id.merge$index\n",
    "    \n",
    "    ## Genotype\n",
    "    Geno <- seqGetData(genofile, \"$dosage\")\n",
    "    Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "    \n",
    "    summary_stat <- NULL\n",
    "    GTSinvG_rare <- NULL\n",
    "    cov_cond <- NULL\n",
    "    \n",
    "    if(!is.null(Geno))\n",
    "    {\n",
    "      # Summary statistics\n",
    "      genotype <- matrix_impute(Geno)\n",
    "      variant_info <- data.frame(chr,pos,ref,alt,row.names=NULL)\n",
    "      \n",
    "      ## Annotation\n",
    "      Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.missense,]\n",
    "      \n",
    "      try(summary_stat <- MetaSTAARlite_worker_sumstat(genotype,obj_nullmodel,variant_info,qc_label,\n",
    "                                                       Anno.Int.PHRED.sub.category),silent=silent)\n",
    "      \n",
    "      # Covariance matrices\n",
    "      genotype <- matrix_flip(Geno)$Geno\n",
    "      genotype <- as(genotype,\"dgCMatrix\")\n",
    "      \n",
    "      rm(Geno)\n",
    "      gc()\n",
    "      \n",
    "      try(GTSinvG_rare <- MetaSTAARlite_worker_cov(genotype,obj_nullmodel,cov_maf_cutoff,\n",
    "                                                   qc_label,signif.digits),silent=silent)\n",
    "      \n",
    "      # Covariance matrices for conditional analysis\n",
    "      if(!is.null(known_loci))\n",
    "      {\n",
    "        try(cov_cond <- MetaSTAARlite_worker_cov_cond(genotype,G_SNV,obj_nullmodel,variant_info,variant_adj_info),silent=silent)\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    summary_stat_list[[\"missense\"]] <- summary_stat\n",
    "    GTSinvG_rare_list[[\"missense\"]] <- GTSinvG_rare\n",
    "    cov_cond_list[[\"missense\"]] <- cov_cond\n",
    "    \n",
    "    #################################################\n",
    "    #         disruptive missense\n",
    "    #################################################\n",
    "    lof.in.dmissense <- (GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\")\n",
    "    variant.id.gene.category <- variant.id.gene[lof.in.dmissense]\n",
    "    \n",
    "    seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "    \n",
    "    pos <- as.integer(seqGetData(genofile, \"position\"))\n",
    "    ref <- as.character(seqGetData(genofile, \"$ref\"))\n",
    "    alt <- as.character(seqGetData(genofile, \"$alt\"))\n",
    "    if(check_qc_label){\n",
    "      qc_label <- as.character(seqGetData(genofile, QC_label))\n",
    "    }else{\n",
    "      qc_label <- NULL\n",
    "    }\n",
    "    \n",
    "    ## genotype id\n",
    "    id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "    # id.genotype.match <- rep(0,length(id.genotype))\n",
    "    \n",
    "    id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "    phenotype.id.merge <- data.frame(phenotype.id)\n",
    "    phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "    id.genotype.match <- phenotype.id.merge$index\n",
    "    \n",
    "    ## Genotype\n",
    "    Geno <- seqGetData(genofile, \"$dosage\")\n",
    "    Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "    \n",
    "    summary_stat <- NULL\n",
    "    GTSinvG_rare <- NULL\n",
    "    cov_cond <- NULL\n",
    "    \n",
    "    if(!is.null(Geno))\n",
    "    {\n",
    "      # Summary statistics\n",
    "      genotype <- matrix_impute(Geno)\n",
    "      variant_info <- data.frame(chr,pos,ref,alt,row.names=NULL)\n",
    "      \n",
    "      ## Annotation\n",
    "      Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.dmissense,]\n",
    "      \n",
    "      try(summary_stat <- MetaSTAARlite_worker_sumstat(genotype,obj_nullmodel,variant_info,qc_label,\n",
    "                                                       Anno.Int.PHRED.sub.category),silent=silent)\n",
    "      \n",
    "      # Covariance matrices\n",
    "      genotype <- matrix_flip(Geno)$Geno\n",
    "      genotype <- as(genotype,\"dgCMatrix\")\n",
    "      \n",
    "      rm(Geno)\n",
    "      gc()\n",
    "      \n",
    "      try(GTSinvG_rare <- MetaSTAARlite_worker_cov(genotype,obj_nullmodel,cov_maf_cutoff,\n",
    "                                                   qc_label,signif.digits),silent=silent)\n",
    "      \n",
    "      # Covariance matrices for conditional analysis\n",
    "      if(!is.null(known_loci))\n",
    "      {\n",
    "        try(cov_cond <- MetaSTAARlite_worker_cov_cond(genotype,G_SNV,obj_nullmodel,variant_info,variant_adj_info),silent=silent)\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    summary_stat_list[[\"disruptive_missense\"]] <- summary_stat\n",
    "    GTSinvG_rare_list[[\"disruptive_missense\"]] <- GTSinvG_rare\n",
    "    cov_cond_list[[\"disruptive_missense\"]] <- cov_cond\n",
    "    \n",
    "    seqResetFilter(genofile)\n",
    "    \n",
    "    if(!is.null(known_loci))\n",
    "    {\n",
    "      return(list(summary_stat_list=summary_stat_list,\n",
    "                  GTSinvG_rare_list=GTSinvG_rare_list,\n",
    "                  cov_cond_list=cov_cond_list))\n",
    "    }else\n",
    "    {\n",
    "      return(list(summary_stat_list=summary_stat_list,\n",
    "                  GTSinvG_rare_list=GTSinvG_rare_list))\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  \n",
    "  ###########################################################\n",
    "  #           User Input\n",
    "  ###########################################################\n",
    "  ## Null model\n",
    "  obj_nullmodel <- get(load(Null_Model))\n",
    "\n",
    "## Parameter\n",
    "QC_label <- \"annotation/filter\"\n",
    "geno_missing_imputation <- \"mean\"\n",
    "variant_type <- \"SNV\"\n",
    "\n",
    "## Annotation_dir\n",
    "Annotation_dir <- \"annotation/info/FunctionalAnnotation\"\n",
    "## Annotation channel\n",
    "Annotation_name_catalog <- read.csv(Annotation_name_catalog)\n",
    "## Use_annotation_weights\n",
    "Use_annotation_weights <- TRUE\n",
    "## Annotation name\n",
    "Annotation_name <- c(\"CADD\",\"LINSIGHT\",\"FATHMM.XF\",\"aPC.EpigeneticActive\",\"aPC.EpigeneticRepressed\",\"aPC.EpigeneticTranscription\",\n",
    "                     \"aPC.Conservation\",\"aPC.LocalDiversity\",\"aPC.Mappability\",\"aPC.TF\",\"aPC.Protein\")\n",
    "\n",
    "  ## output file name\n",
    "  output_file_name <- paste0(OUTPUT_PATH,\"/AoU_TC_Coding\")\n",
    "  \n",
    "  ###########################################################\n",
    "  #           Main Function \n",
    "  ###########################################################\n",
    "  ## gene number in job\n",
    "  gene_num_in_array <- 50\n",
    "  group.num.allchr <- ceiling(table(genes_info[,2])/gene_num_in_array)\n",
    "  sum(group.num.allchr)\n",
    "  \n",
    "  chr <- which.max(arrayid <= cumsum(group.num.allchr))\n",
    "  group.num <- group.num.allchr[chr]\n",
    "  \n",
    "  if (chr == 1){\n",
    "    groupid <- arrayid\n",
    "  }else{\n",
    "    groupid <- arrayid - cumsum(group.num.allchr)[chr-1]\n",
    "  }\n",
    "  \n",
    "  genes_info_chr <- genes_info[genes_info[,2]==chr,]\n",
    "  sub_seq_num <- dim(genes_info_chr)[1]\n",
    "  \n",
    "  if(groupid < group.num){\n",
    "    sub_seq_id <- ((groupid - 1)*gene_num_in_array + 1):(groupid*gene_num_in_array)\n",
    "  }else{\n",
    "    sub_seq_id <- ((groupid - 1)*gene_num_in_array + 1):sub_seq_num\n",
    "  }\n",
    "  \n",
    "  ## exclude large coding masks\n",
    "  if(arrayid==57){\n",
    "    sub_seq_id <- setdiff(sub_seq_id,840)\n",
    "  }\n",
    "  \n",
    "  if(arrayid==112){\n",
    "    sub_seq_id <- setdiff(sub_seq_id,c(543,544))\n",
    "  }\n",
    "  \n",
    "  if(arrayid==113){\n",
    "    sub_seq_id <- setdiff(sub_seq_id,c(575,576,577,578,579,580,582))\n",
    "  }\n",
    "  \n",
    "  ## aGDS file\n",
    "  agds.path <- GDS_File\n",
    "\n",
    "  genofile <- seqOpen(agds.path)\n",
    "  \n",
    "  genes <- genes_info\n",
    "  \n",
    "  coding_sumstat <- list()\n",
    "  coding_cov <- list()\n",
    "  for(kk in sub_seq_id){\n",
    "    print(kk)\n",
    "    gene_name <- genes_info_chr[kk,1]\n",
    "    results_temp <- coding_MetaSTAARlite_worker(chr=chr,gene_name=gene_name,genofile=genofile,obj_nullmodel=obj_nullmodel,genes=genes,\n",
    "                                                cov_maf_cutoff=0.05,signif.digits=NULL,\n",
    "                                                QC_label=QC_label,check_qc_label=TRUE,variant_type=variant_type,\n",
    "                                                Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                                Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name)\n",
    "    coding_sumstat[[gene_name]] <- results_temp$summary_stat_list\n",
    "    coding_cov[[gene_name]] <- results_temp$GTSinvG_rare_list\n",
    "  }\n",
    "  \n",
    "  save(coding_sumstat,file=paste0(output_file_name,\"_sumstat_\",arrayid,\".Rdata\"),compress = \"xz\")\n",
    "  save(coding_cov,file=paste0(output_file_name,\"_cov_\",arrayid,\".Rdata\"),compress = \"xz\")\n",
    "  \n",
    "  seqClose(genofile)\n",
    "  \n",
    "})[3]\n",
    "\n",
    "save(time,file = paste0(output_file_name,\"_time_\",arrayid,\".Rdata\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Meta_Coding_AoU.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile Meta_Coding_AoU.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${arrayid} ${Null_Model} ${Annotation_name_catalog} ${GDS_File} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for(arrayid in c(112,113,332)){\n",
    "        if(arrayid < 41){\n",
    "            chr <- 1\n",
    "        }else if(arrayid < 66){\n",
    "            chr <- 2\n",
    "        }else if(arrayid < 87){\n",
    "            chr <- 3\n",
    "        }else if(arrayid < 102){\n",
    "            chr <- 4\n",
    "        }else if(arrayid < 120){\n",
    "            chr <- 5\n",
    "        }else if(arrayid < 141){\n",
    "            chr <- 6\n",
    "        }else if(arrayid < 159){\n",
    "            chr <- 7\n",
    "        }else if(arrayid < 173){\n",
    "            chr <- 8\n",
    "        }else if(arrayid < 189){\n",
    "            chr <- 9\n",
    "        }else if(arrayid < 204){\n",
    "            chr <- 10\n",
    "        }else if(arrayid < 230){\n",
    "            chr <- 11\n",
    "        }else if(arrayid < 250){\n",
    "            chr <- 12\n",
    "        }else if(arrayid < 257){\n",
    "            chr <- 13\n",
    "        }else if(arrayid < 269){\n",
    "            chr <- 14\n",
    "        }else if(arrayid < 281){\n",
    "            chr <- 15\n",
    "        }else if(arrayid < 298){\n",
    "            chr <- 16\n",
    "        }else if(arrayid < 321){\n",
    "            chr <- 17\n",
    "        }else if(arrayid < 327){\n",
    "            chr <- 18\n",
    "        }else if(arrayid < 355){\n",
    "            chr <- 19\n",
    "        }else if(arrayid < 366){\n",
    "            chr <- 20\n",
    "        }else if(arrayid < 371){\n",
    "            chr <- 21\n",
    "        }else{\n",
    "            chr <- 22\n",
    "        }\n",
    "        \n",
    "        tasks <- rbind(tasks, data.frame(\n",
    "            '--env arrayid'=arrayid,\n",
    "            '--input Null_Model'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/MetaSTAAR/TC_Null_Model.RData\",\n",
    "            '--input Annotation_name_catalog'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataGDS/acaf_threshold_v7/Annotation_name_catalog.csv\",\n",
    "            '--input GDS_File'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataAGDS/exome_v7.1/exome.chr\",chr,\".gds\"),\n",
    "            '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/MetaSTAAR/Meta_Coding_AoU.R\",\n",
    "            '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/MetaSTAAR/AoU_Sumstats/\",\n",
    "            check.names = FALSE\n",
    "        )) \n",
    "    }\n",
    "  \n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://Meta_Coding_AoU.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][ 28.4 KiB/ 28.4 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/28.4 KiB.                                     \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp Meta_Coding_AoU.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/MetaSTAAR/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: meta-codin--williamsjacr--240911-212807-78\n",
      "  job-name: meta-coding-aou\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/9957613312611743355\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/6784081199623067344\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3520714611431293840\n",
      "Launched job-id: meta-codin--williamsjacr--240911-212807-78\n",
      "3 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'meta-codin--williamsjacr--240911-212807-78' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'meta-codin--williamsjacr--240911-212807-78' --users 'williamsjacr'\n",
      "Waiting for job to complete...\n",
      "Monitoring for failed tasks to retry...\n",
      "*** This dsub process must continue running to retry failed tasks.\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 64 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script Meta_Coding_AoU.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "code_folding": [
     39,
     113,
     189
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Meta_Coding_AoU_LongMasks.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile Meta_Coding_AoU_LongMasks.R\n",
    "rm(list=ls())\n",
    "gc()\n",
    "\n",
    "arrayid_longmask <- as.numeric(commandArgs(TRUE)[1])\n",
    "\n",
    "Null_Model <- commandArgs(TRUE)[2]\n",
    "print(Null_Model)\n",
    "\n",
    "Annotation_name_catalog <- commandArgs(TRUE)[3]\n",
    "print(Annotation_name_catalog)\n",
    "\n",
    "GDS_File_Chr2 <- commandArgs(TRUE)[4]\n",
    "print(GDS_File_Chr2)\n",
    "\n",
    "GDS_File_Chr5 <- commandArgs(TRUE)[5]\n",
    "print(GDS_File_Chr5)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[6]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "time <- system.time({\n",
    "  ## load required packages\n",
    "  library(gdsfmt)\n",
    "  library(SeqArray)\n",
    "  library(SeqVarTools)\n",
    "  library(STAAR)\n",
    "  library(STAARpipeline)\n",
    "  library(readr)\n",
    "  library(dplyr)\n",
    "  library(stringr)\n",
    "  \n",
    "  library(devtools)\n",
    "#  build(R_PACKAGE,path = \"\")\n",
    "#  install.packages(\"MetaSTAAR_0.9.6.3.tar.gz\", repos = NULL, type=\"source\")\n",
    "  library(MetaSTAAR)\n",
    "  library(expm)\n",
    "  library(Matrix)\n",
    "  \n",
    "  MetaSTAARlite_worker_sumstat <- function(genotype,obj_nullmodel,variant_info,qc_label=NULL,\n",
    "                                           annotation_phred=NULL,for_individual_analysis=FALSE){\n",
    "    \n",
    "    if(!inherits(genotype, \"matrix\") && !inherits(genotype, \"Matrix\")){\n",
    "      stop(\"genotype is not a matrix!\")\n",
    "    }\n",
    "    \n",
    "    if(inherits(genotype, \"sparseMatrix\")){\n",
    "      genotype <- as.matrix(genotype)\n",
    "    }\n",
    "    \n",
    "    if(dim(genotype)[2] != dim(variant_info)[1]){\n",
    "      stop(paste0(\"Dimensions don't match for genotype and variant_info!\"))\n",
    "    }\n",
    "    \n",
    "    if(!is.null(qc_label) && dim(variant_info)[1] != length(qc_label)){\n",
    "      stop(paste0(\"Dimensions don't match for variant_info and qc_label!\"))\n",
    "    }\n",
    "    \n",
    "    if(!is.null(annotation_phred) && dim(annotation_phred)[1] != dim(variant_info)[1]){\n",
    "      stop(paste0(\"Dimensions don't match for annotation_phred and variant_info!\"))\n",
    "    }\n",
    "    \n",
    "    N <- dim(genotype)[1]\n",
    "    alt_AC <- as.integer(colSums(2 - genotype))\n",
    "    MAC <- as.integer(pmin(alt_AC, 2 * N - alt_AC))\n",
    "    genotype <- matrix_flip(genotype)\n",
    "    MAF <- genotype$MAF\n",
    "    variant_label <- as.vector(MAF>0)\n",
    "    Geno <- genotype$Geno[,variant_label,drop=FALSE]\n",
    "    Geno <- as(Geno,\"dgCMatrix\")\n",
    "    rm(genotype)\n",
    "    gc()\n",
    "    \n",
    "    if(obj_nullmodel$relatedness){\n",
    "      if(!obj_nullmodel$sparse_kins){\n",
    "        stop(paste0(\"Please use a sparse genetic relatedness matrix when fitting the null model!\"))\n",
    "      }\n",
    "    }else{\n",
    "      if(obj_nullmodel$family[1] == \"binomial\"){\n",
    "        obj_nullmodel$Sigma_i <- Diagonal(x = obj_nullmodel$weights)\n",
    "      }else if(obj_nullmodel$family[1] == \"gaussian\"){\n",
    "        obj_nullmodel$Sigma_i <- Diagonal(length(obj_nullmodel$y)) / summary(obj_nullmodel)$dispersion\n",
    "      }\n",
    "      obj_nullmodel$Sigma_iX <- obj_nullmodel$Sigma_i %*% model.matrix(obj_nullmodel)\n",
    "      obj_nullmodel$cov <- solve(t(model.matrix(obj_nullmodel)) %*% obj_nullmodel$Sigma_i %*% model.matrix(obj_nullmodel))\n",
    "      obj_nullmodel$scaled.residuals <- (obj_nullmodel$y - obj_nullmodel$fitted.values) / summary(obj_nullmodel)$dispersion\n",
    "    }\n",
    "    \n",
    "    GTSinvX_cov <- as(t(Geno) %*% obj_nullmodel$Sigma_iX,\"matrix\") %*% sqrtm(obj_nullmodel$cov)\n",
    "    #V <- diag(t(Geno) %*% obj_nullmodel$Sigma_i %*% Geno - GTSinvX_cov %*% t(GTSinvX_cov))\n",
    "    V <- colSums(Geno * (obj_nullmodel$Sigma_i %*% Geno)) - rowSums(GTSinvX_cov^2) # faster for large number of variants\n",
    "    U <- as.vector(t(Geno) %*% obj_nullmodel$scaled.residuals)\n",
    "    rm(Geno)\n",
    "    gc()\n",
    "    \n",
    "    if (!is.null(qc_label)){\n",
    "      sumstat <- data.frame(variant_info[,c(\"chr\",\"pos\",\"ref\",\"alt\")],qc_label,alt_AC,MAC,MAF,N)\n",
    "    }else{\n",
    "      sumstat <- data.frame(variant_info[,c(\"chr\",\"pos\",\"ref\",\"alt\")],alt_AC,MAC,MAF,N)\n",
    "    }\n",
    "    if (!is.null(annotation_phred)){\n",
    "      sumstat <- cbind(sumstat,annotation_phred)\n",
    "    }\n",
    "    sumstat <- sumstat[variant_label,]\n",
    "    if(!for_individual_analysis){\n",
    "      sumstat <- cbind(sumstat,U,V,GTSinvX_cov)\n",
    "    }else{\n",
    "      sumstat <- cbind(sumstat,U,V)\n",
    "    }\n",
    "    \n",
    "    return(sumstat)\n",
    "  }\n",
    "  \n",
    "  MetaSTAARlite_worker_cov <- function(genotype,obj_nullmodel,cov_maf_cutoff=0.05,\n",
    "                                       qc_label=NULL,signif.digits=3){\n",
    "    \n",
    "    if(!inherits(genotype, \"matrix\") && !inherits(genotype, \"Matrix\")){\n",
    "      stop(\"genotype is not a matrix!\")\n",
    "    }\n",
    "    \n",
    "    if(cov_maf_cutoff < 0 | cov_maf_cutoff > 0.5){\n",
    "      stop(\"cov_maf_cutoff should be a number between 0 and 0.5!\")\n",
    "    }\n",
    "    \n",
    "    if(cov_maf_cutoff == 0.5){\n",
    "      cov_maf_cutoff <- 0.5 + 1e-16\n",
    "    }\n",
    "    \n",
    "    if(inherits(genotype, \"sparseMatrix\")){\n",
    "      MAF <- colMeans(genotype)/2\n",
    "      if (!is.null(qc_label)){\n",
    "        RV_label <- as.vector((MAF<cov_maf_cutoff)&(MAF>0)&(qc_label==\"PASS\"))\n",
    "      }else{\n",
    "        RV_label <- as.vector((MAF<cov_maf_cutoff)&(MAF>0))\n",
    "      }\n",
    "      Geno_rare <- genotype[,RV_label,drop=FALSE]\n",
    "      Geno_rare <- as(Geno_rare,\"dgCMatrix\")\n",
    "    }else{\n",
    "      genotype <- matrix_flip(genotype)\n",
    "      MAF <- genotype$MAF\n",
    "      if (!is.null(qc_label)){\n",
    "        RV_label <- as.vector((MAF<cov_maf_cutoff)&(MAF>0)&(qc_label==\"PASS\"))\n",
    "      }else{\n",
    "        RV_label <- as.vector((MAF<cov_maf_cutoff)&(MAF>0))\n",
    "      }\n",
    "      Geno_rare <- genotype$Geno[,RV_label,drop=FALSE]\n",
    "      Geno_rare <- as(Geno_rare,\"dgCMatrix\")\n",
    "    }\n",
    "    \n",
    "    rm(genotype,MAF)\n",
    "    gc()\n",
    "    \n",
    "    if(obj_nullmodel$relatedness){\n",
    "      if(!obj_nullmodel$sparse_kins){\n",
    "        stop(paste0(\"Please use a sparse genetic relatedness matrix when fitting the null model!\"))\n",
    "      }\n",
    "    }else{\n",
    "      if(obj_nullmodel$family[1] == \"binomial\"){\n",
    "        obj_nullmodel$Sigma_i <- Diagonal(x = obj_nullmodel$weights)\n",
    "      }else if(obj_nullmodel$family[1] == \"gaussian\"){\n",
    "        obj_nullmodel$Sigma_i <- Diagonal(length(obj_nullmodel$y)) / summary(obj_nullmodel)$dispersion\n",
    "      }\n",
    "      obj_nullmodel$Sigma_iX <- obj_nullmodel$Sigma_i %*% model.matrix(obj_nullmodel)\n",
    "      obj_nullmodel$cov <- solve(t(model.matrix(obj_nullmodel)) %*% obj_nullmodel$Sigma_i %*% model.matrix(obj_nullmodel))\n",
    "      obj_nullmodel$scaled.residuals <- (obj_nullmodel$y - obj_nullmodel$fitted.values) / summary(obj_nullmodel)$dispersion\n",
    "    }\n",
    "    \n",
    "    GTSinvG_rare <- t(obj_nullmodel$Sigma_i %*% Geno_rare) %*% Geno_rare\n",
    "    rm(Geno_rare)\n",
    "    gc()\n",
    "    \n",
    "    GTSinvG_rare <- as(GTSinvG_rare,\"TsparseMatrix\")\n",
    "    remove_ind <- (GTSinvG_rare@j < GTSinvG_rare@i) # Be careful about multi-allelic issue\n",
    "    GTSinvG_rare@i <- GTSinvG_rare@i[!remove_ind]\n",
    "    GTSinvG_rare@j <- GTSinvG_rare@j[!remove_ind]\n",
    "    GTSinvG_rare@x <- GTSinvG_rare@x[!remove_ind]\n",
    "    rm(remove_ind)\n",
    "    gc()\n",
    "    GTSinvG_rare <- as(GTSinvG_rare,\"CsparseMatrix\")\n",
    "    \n",
    "    ### Create a version of GTSinvG_rare with rounded significant digits\n",
    "    if(!is.null(signif.digits)){\n",
    "      GTSinvG_rare <- signif(GTSinvG_rare, digits = signif.digits)\n",
    "    }\n",
    "    \n",
    "    return(GTSinvG_rare)\n",
    "  }\n",
    "  \n",
    "  \n",
    "  coding_MetaSTAARlite_worker <- function(chr,gene_name,genofile,obj_nullmodel,genes,known_loci=NULL,\n",
    "                                          cov_maf_cutoff=0.05,signif.digits=NULL,\n",
    "                                          QC_label=\"annotation/filter\",check_qc_label=FALSE,variant_type=c(\"SNV\",\"Indel\",\"variant\"),\n",
    "                                          Annotation_dir=\"annotation/info/FunctionalAnnotation\",Annotation_name_catalog,\n",
    "                                          Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,\n",
    "                                          silent=FALSE){\n",
    "    \n",
    "    ## evaluate choices\n",
    "    variant_type <- match.arg(variant_type)\n",
    "    \n",
    "    phenotype.id <- obj_nullmodel$id_include\n",
    "    \n",
    "    filter <- seqGetData(genofile, QC_label)\n",
    "    if(variant_type==\"variant\")\n",
    "    {\n",
    "      if(check_qc_label)\n",
    "      {\n",
    "        SNVlist <- TRUE\n",
    "      }else\n",
    "      {\n",
    "        SNVlist <- filter == \"PASS\"\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    if(variant_type==\"SNV\")\n",
    "    {\n",
    "      if(check_qc_label)\n",
    "      {\n",
    "        SNVlist <- isSNV(genofile)\n",
    "      }else\n",
    "      {\n",
    "        SNVlist <- (filter == \"PASS\") & isSNV(genofile)\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    if(variant_type==\"Indel\")\n",
    "    {\n",
    "      if(check_qc_label)\n",
    "      {\n",
    "        SNVlist <- !isSNV(genofile)\n",
    "      }else\n",
    "      {\n",
    "        SNVlist <- (filter == \"PASS\") & (!isSNV(genofile))\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    position <- as.numeric(seqGetData(genofile, \"position\"))\n",
    "    variant.id <- seqGetData(genofile, \"variant.id\")\n",
    "    \n",
    "    if(!is.null(known_loci))\n",
    "    {\n",
    "      allele <- seqGetData(genofile, \"allele\")\n",
    "      \n",
    "      loc_SNV <- c()\n",
    "      for (i in 1:dim(known_loci)[1]){\n",
    "        loc_SNV <- c(loc_SNV,which((position==known_loci$POS[i])&(allele==paste0(known_loci$REF[i],\",\",known_loci$ALT[i]))))\n",
    "      }\n",
    "      \n",
    "      seqSetFilter(genofile,variant.id=variant.id[loc_SNV],sample.id=phenotype.id)\n",
    "      \n",
    "      G_SNV <- seqGetData(genofile, \"$dosage\")\n",
    "      \n",
    "      if (!is.null(G_SNV)){\n",
    "        id.SNV <- seqGetData(genofile,\"sample.id\")\n",
    "        id.SNV.match <- rep(0,length(id.SNV))\n",
    "        \n",
    "        for(i in 1:length(id.SNV))\n",
    "        {\n",
    "          id.SNV.match[i] <- which.max(id.SNV==phenotype.id[i])\n",
    "        }\n",
    "        \n",
    "        G_SNV <- G_SNV[id.SNV.match,,drop=FALSE]\n",
    "        G_SNV_MAF <- apply(G_SNV, 2, function(x){sum(x[!is.na(x)])/2/sum(!is.na(x))})\n",
    "        for (i in 1:length(G_SNV_MAF)){\n",
    "          if (G_SNV_MAF[i] <= 0.5){\n",
    "            G_SNV[is.na(G_SNV[,i]),i] <- 0\n",
    "          }\n",
    "          else{\n",
    "            G_SNV[is.na(G_SNV[,i]),i] <- 2\n",
    "          }\n",
    "        }\n",
    "        \n",
    "        pos_adj <- as.integer(seqGetData(genofile, \"position\"))\n",
    "        ref_adj <- as.character(seqGetData(genofile, \"$ref\"))\n",
    "        alt_adj <- as.character(seqGetData(genofile, \"$alt\"))\n",
    "        variant_adj_info <- data.frame(chr,pos_adj,ref_adj,alt_adj)\n",
    "        colnames(variant_adj_info) <- c(\"chr\",\"pos\",\"ref\",\"alt\")\n",
    "        variant_adj_info\n",
    "        \n",
    "        seqResetFilter(genofile)\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    rm(filter)\n",
    "    gc()\n",
    "    \n",
    "    ### Gene\n",
    "    kk <- which(genes[,1]==gene_name)\n",
    "    \n",
    "    sub_start_loc <- genes[kk,3]\n",
    "    sub_end_loc <- genes[kk,4]\n",
    "    \n",
    "    is.in <- (SNVlist)&(position>=sub_start_loc)&(position<=sub_end_loc)\n",
    "    variant.id.gene <- variant.id[is.in]\n",
    "    \n",
    "    rm(position)\n",
    "    gc()\n",
    "    \n",
    "    seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "    \n",
    "    ## Gencode_Exonic\n",
    "    GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "    ## Gencode\n",
    "    GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "    ## Meta.SVM.Pred\n",
    "    MetaSVM_pred <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"MetaSVM\")]))\n",
    "    \n",
    "    ################################################\n",
    "    #           Coding\n",
    "    ################################################\n",
    "    variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "    lof.in.coding <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")|(GENCODE.EXONIC.Category==\"nonsynonymous SNV\")|(GENCODE.EXONIC.Category==\"synonymous SNV\")\n",
    "    variant.id.gene <- variant.id.gene[lof.in.coding]\n",
    "    \n",
    "    seqSetFilter(genofile,variant.id=variant.id.gene,sample.id=phenotype.id)\n",
    "    \n",
    "    ## Gencode_Exonic\n",
    "    GENCODE.EXONIC.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.EXONIC.Category\")]))\n",
    "    ## Gencode\n",
    "    GENCODE.Category <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"GENCODE.Category\")]))\n",
    "    ## Meta.SVM.Pred\n",
    "    MetaSVM_pred <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==\"MetaSVM\")]))\n",
    "    \n",
    "    ## Annotation\n",
    "    Anno.Int.PHRED.sub <- NULL\n",
    "    Anno.Int.PHRED.sub.name <- NULL\n",
    "    \n",
    "    if(variant_type==\"SNV\")\n",
    "    {\n",
    "      if(Use_annotation_weights)\n",
    "      {\n",
    "        for(k in 1:length(Annotation_name))\n",
    "        {\n",
    "          if(Annotation_name[k]%in%Annotation_name_catalog$name)\n",
    "          {\n",
    "            Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "            Annotation.PHRED <- seqGetData(genofile, paste0(Annotation_dir,Annotation_name_catalog$dir[which(Annotation_name_catalog$name==Annotation_name[k])]))\n",
    "            \n",
    "            if(Annotation_name[k]==\"CADD\")\n",
    "            {\n",
    "              Annotation.PHRED[is.na(Annotation.PHRED)] <- 0\n",
    "            }\n",
    "            Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "          }\n",
    "        }\n",
    "        \n",
    "        Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "        colnames(Anno.Int.PHRED.sub) <- Anno.Int.PHRED.sub.name\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    ################################################\n",
    "    #                  plof_ds\n",
    "    ################################################\n",
    "    variant.id.gene <- seqGetData(genofile, \"variant.id\")\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")|((GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\"))\n",
    "    variant.id.gene.category <- variant.id.gene[lof.in.plof]\n",
    "    \n",
    "    seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "    \n",
    "    pos <- as.integer(seqGetData(genofile, \"position\"))\n",
    "    ref <- as.character(seqGetData(genofile, \"$ref\"))\n",
    "    alt <- as.character(seqGetData(genofile, \"$alt\"))\n",
    "    if(check_qc_label){\n",
    "      qc_label <- as.character(seqGetData(genofile, QC_label))\n",
    "    }else{\n",
    "      qc_label <- NULL\n",
    "    }\n",
    "    \n",
    "    ## genotype id\n",
    "    id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "    # id.genotype.match <- rep(0,length(id.genotype))\n",
    "    \n",
    "    id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "    phenotype.id.merge <- data.frame(phenotype.id)\n",
    "    phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "    id.genotype.match <- phenotype.id.merge$index\n",
    "    \n",
    "    ## Genotype\n",
    "    Geno <- seqGetData(genofile, \"$dosage\")\n",
    "    Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "    \n",
    "    summary_stat_list <- list()\n",
    "    GTSinvG_rare_list <- list()\n",
    "    cov_cond_list <- list()\n",
    "    \n",
    "    summary_stat <- NULL\n",
    "    GTSinvG_rare <- NULL\n",
    "    cov_cond <- NULL\n",
    "    \n",
    "    if(!is.null(Geno))\n",
    "    {\n",
    "      # Summary statistics\n",
    "      genotype <- matrix_impute(Geno)\n",
    "      variant_info <- data.frame(chr,pos,ref,alt,row.names=NULL)\n",
    "      \n",
    "      ## Annotation\n",
    "      Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.plof,]\n",
    "      \n",
    "      try(summary_stat <- MetaSTAARlite_worker_sumstat(genotype,obj_nullmodel,variant_info,qc_label,\n",
    "                                                       Anno.Int.PHRED.sub.category),silent=silent)\n",
    "      \n",
    "      # Covariance matrices\n",
    "      genotype <- matrix_flip(Geno)$Geno\n",
    "      genotype <- as(genotype,\"dgCMatrix\")\n",
    "      \n",
    "      rm(Geno)\n",
    "      gc()\n",
    "      \n",
    "      try(GTSinvG_rare <- MetaSTAARlite_worker_cov(genotype,obj_nullmodel,cov_maf_cutoff,\n",
    "                                                   qc_label,signif.digits),silent=silent)\n",
    "      \n",
    "      # Covariance matrices for conditional analysis\n",
    "      if(!is.null(known_loci))\n",
    "      {\n",
    "        try(cov_cond <- MetaSTAARlite_worker_cov_cond(genotype,G_SNV,obj_nullmodel,variant_info,variant_adj_info),silent=silent)\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    summary_stat_list[[\"plof_ds\"]] <- summary_stat\n",
    "    GTSinvG_rare_list[[\"plof_ds\"]] <- GTSinvG_rare\n",
    "    cov_cond_list[[\"plof_ds\"]] <- cov_cond\n",
    "    \n",
    "    #####################################################\n",
    "    #                      plof\n",
    "    #####################################################\n",
    "    lof.in.plof <- (GENCODE.EXONIC.Category==\"stopgain\")|(GENCODE.EXONIC.Category==\"stoploss\")|(GENCODE.Category==\"splicing\")|(GENCODE.Category==\"exonic;splicing\")|(GENCODE.Category==\"ncRNA_splicing\")|(GENCODE.Category==\"ncRNA_exonic;splicing\")\n",
    "    variant.id.gene.category <- variant.id.gene[lof.in.plof]\n",
    "    \n",
    "    seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "    \n",
    "    pos <- as.integer(seqGetData(genofile, \"position\"))\n",
    "    ref <- as.character(seqGetData(genofile, \"$ref\"))\n",
    "    alt <- as.character(seqGetData(genofile, \"$alt\"))\n",
    "    if(check_qc_label){\n",
    "      qc_label <- as.character(seqGetData(genofile, QC_label))\n",
    "    }else{\n",
    "      qc_label <- NULL\n",
    "    }\n",
    "    \n",
    "    ## genotype id\n",
    "    id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "    # id.genotype.match <- rep(0,length(id.genotype))\n",
    "    \n",
    "    id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "    phenotype.id.merge <- data.frame(phenotype.id)\n",
    "    phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "    id.genotype.match <- phenotype.id.merge$index\n",
    "    \n",
    "    ## Genotype\n",
    "    Geno <- seqGetData(genofile, \"$dosage\")\n",
    "    Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "    \n",
    "    summary_stat <- NULL\n",
    "    GTSinvG_rare <- NULL\n",
    "    cov_cond <- NULL\n",
    "    \n",
    "    if(!is.null(Geno))\n",
    "    {\n",
    "      # Summary statistics\n",
    "      genotype <- matrix_impute(Geno)\n",
    "      variant_info <- data.frame(chr,pos,ref,alt,row.names=NULL)\n",
    "      \n",
    "      ## Annotation\n",
    "      Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.plof,]\n",
    "      \n",
    "      try(summary_stat <- MetaSTAARlite_worker_sumstat(genotype,obj_nullmodel,variant_info,qc_label,\n",
    "                                                       Anno.Int.PHRED.sub.category),silent=silent)\n",
    "      \n",
    "      # Covariance matrices\n",
    "      genotype <- matrix_flip(Geno)$Geno\n",
    "      genotype <- as(genotype,\"dgCMatrix\")\n",
    "      \n",
    "      rm(Geno)\n",
    "      gc()\n",
    "      \n",
    "      try(GTSinvG_rare <- MetaSTAARlite_worker_cov(genotype,obj_nullmodel,cov_maf_cutoff,\n",
    "                                                   qc_label,signif.digits),silent=silent)\n",
    "      \n",
    "      # Covariance matrices for conditional analysis\n",
    "      if(!is.null(known_loci))\n",
    "      {\n",
    "        try(cov_cond <- MetaSTAARlite_worker_cov_cond(genotype,G_SNV,obj_nullmodel,variant_info,variant_adj_info),silent=silent)\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    summary_stat_list[[\"plof\"]] <- summary_stat\n",
    "    GTSinvG_rare_list[[\"plof\"]] <- GTSinvG_rare\n",
    "    cov_cond_list[[\"plof\"]] <- cov_cond\n",
    "    \n",
    "    #############################################\n",
    "    #             synonymous\n",
    "    #############################################\n",
    "    lof.in.synonymous <- (GENCODE.EXONIC.Category==\"synonymous SNV\")\n",
    "    variant.id.gene.category <- variant.id.gene[lof.in.synonymous]\n",
    "    \n",
    "    seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "    \n",
    "    pos <- as.integer(seqGetData(genofile, \"position\"))\n",
    "    ref <- as.character(seqGetData(genofile, \"$ref\"))\n",
    "    alt <- as.character(seqGetData(genofile, \"$alt\"))\n",
    "    if(check_qc_label){\n",
    "      qc_label <- as.character(seqGetData(genofile, QC_label))\n",
    "    }else{\n",
    "      qc_label <- NULL\n",
    "    }\n",
    "    \n",
    "    ## genotype id\n",
    "    id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "    # id.genotype.match <- rep(0,length(id.genotype))\n",
    "    \n",
    "    id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "    phenotype.id.merge <- data.frame(phenotype.id)\n",
    "    phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "    id.genotype.match <- phenotype.id.merge$index\n",
    "    \n",
    "    ## Genotype\n",
    "    Geno <- seqGetData(genofile, \"$dosage\")\n",
    "    Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "    \n",
    "    summary_stat <- NULL\n",
    "    GTSinvG_rare <- NULL\n",
    "    cov_cond <- NULL\n",
    "    \n",
    "    if(!is.null(Geno))\n",
    "    {\n",
    "      # Summary statistics\n",
    "      genotype <- matrix_impute(Geno)\n",
    "      variant_info <- data.frame(chr,pos,ref,alt,row.names=NULL)\n",
    "      \n",
    "      ## Annotation\n",
    "      Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.synonymous,]\n",
    "      \n",
    "      try(summary_stat <- MetaSTAARlite_worker_sumstat(genotype,obj_nullmodel,variant_info,qc_label,\n",
    "                                                       Anno.Int.PHRED.sub.category),silent=silent)\n",
    "      \n",
    "      # Covariance matrices\n",
    "      genotype <- matrix_flip(Geno)$Geno\n",
    "      genotype <- as(genotype,\"dgCMatrix\")\n",
    "      \n",
    "      rm(Geno)\n",
    "      gc()\n",
    "      \n",
    "      try(GTSinvG_rare <- MetaSTAARlite_worker_cov(genotype,obj_nullmodel,cov_maf_cutoff,\n",
    "                                                   qc_label,signif.digits),silent=silent)\n",
    "      \n",
    "      # Covariance matrices for conditional analysis\n",
    "      if(!is.null(known_loci))\n",
    "      {\n",
    "        try(cov_cond <- MetaSTAARlite_worker_cov_cond(genotype,G_SNV,obj_nullmodel,variant_info,variant_adj_info),silent=silent)\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    summary_stat_list[[\"synonymous\"]] <- summary_stat\n",
    "    GTSinvG_rare_list[[\"synonymous\"]] <- GTSinvG_rare\n",
    "    cov_cond_list[[\"synonymous\"]] <- cov_cond\n",
    "    \n",
    "    #################################################\n",
    "    #        missense\n",
    "    #################################################\n",
    "    lof.in.missense <- (GENCODE.EXONIC.Category==\"nonsynonymous SNV\")\n",
    "    variant.id.gene.category <- variant.id.gene[lof.in.missense]\n",
    "    \n",
    "    seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "    \n",
    "    pos <- as.integer(seqGetData(genofile, \"position\"))\n",
    "    ref <- as.character(seqGetData(genofile, \"$ref\"))\n",
    "    alt <- as.character(seqGetData(genofile, \"$alt\"))\n",
    "    if(check_qc_label){\n",
    "      qc_label <- as.character(seqGetData(genofile, QC_label))\n",
    "    }else{\n",
    "      qc_label <- NULL\n",
    "    }\n",
    "    \n",
    "    ## genotype id\n",
    "    id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "    # id.genotype.match <- rep(0,length(id.genotype))\n",
    "    \n",
    "    id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "    phenotype.id.merge <- data.frame(phenotype.id)\n",
    "    phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "    id.genotype.match <- phenotype.id.merge$index\n",
    "    \n",
    "    ## Genotype\n",
    "    Geno <- seqGetData(genofile, \"$dosage\")\n",
    "    Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "    \n",
    "    summary_stat <- NULL\n",
    "    GTSinvG_rare <- NULL\n",
    "    cov_cond <- NULL\n",
    "    \n",
    "    if(!is.null(Geno))\n",
    "    {\n",
    "      # Summary statistics\n",
    "      genotype <- matrix_impute(Geno)\n",
    "      variant_info <- data.frame(chr,pos,ref,alt,row.names=NULL)\n",
    "      \n",
    "      ## Annotation\n",
    "      Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.missense,]\n",
    "      \n",
    "      try(summary_stat <- MetaSTAARlite_worker_sumstat(genotype,obj_nullmodel,variant_info,qc_label,\n",
    "                                                       Anno.Int.PHRED.sub.category),silent=silent)\n",
    "      \n",
    "      # Covariance matrices\n",
    "      genotype <- matrix_flip(Geno)$Geno\n",
    "      genotype <- as(genotype,\"dgCMatrix\")\n",
    "      \n",
    "      rm(Geno)\n",
    "      gc()\n",
    "      \n",
    "      try(GTSinvG_rare <- MetaSTAARlite_worker_cov(genotype,obj_nullmodel,cov_maf_cutoff,\n",
    "                                                   qc_label,signif.digits),silent=silent)\n",
    "      \n",
    "      # Covariance matrices for conditional analysis\n",
    "      if(!is.null(known_loci))\n",
    "      {\n",
    "        try(cov_cond <- MetaSTAARlite_worker_cov_cond(genotype,G_SNV,obj_nullmodel,variant_info,variant_adj_info),silent=silent)\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    summary_stat_list[[\"missense\"]] <- summary_stat\n",
    "    GTSinvG_rare_list[[\"missense\"]] <- GTSinvG_rare\n",
    "    cov_cond_list[[\"missense\"]] <- cov_cond\n",
    "    \n",
    "    #################################################\n",
    "    #         disruptive missense\n",
    "    #################################################\n",
    "    lof.in.dmissense <- (GENCODE.EXONIC.Category==\"nonsynonymous SNV\")&(MetaSVM_pred==\"D\")\n",
    "    variant.id.gene.category <- variant.id.gene[lof.in.dmissense]\n",
    "    \n",
    "    seqSetFilter(genofile,variant.id=variant.id.gene.category,sample.id=phenotype.id)\n",
    "    \n",
    "    pos <- as.integer(seqGetData(genofile, \"position\"))\n",
    "    ref <- as.character(seqGetData(genofile, \"$ref\"))\n",
    "    alt <- as.character(seqGetData(genofile, \"$alt\"))\n",
    "    if(check_qc_label){\n",
    "      qc_label <- as.character(seqGetData(genofile, QC_label))\n",
    "    }else{\n",
    "      qc_label <- NULL\n",
    "    }\n",
    "    \n",
    "    ## genotype id\n",
    "    id.genotype <- seqGetData(genofile,\"sample.id\")\n",
    "    # id.genotype.match <- rep(0,length(id.genotype))\n",
    "    \n",
    "    id.genotype.merge <- data.frame(id.genotype,index=seq(1,length(id.genotype)))\n",
    "    phenotype.id.merge <- data.frame(phenotype.id)\n",
    "    phenotype.id.merge <- dplyr::left_join(phenotype.id.merge,id.genotype.merge,by=c(\"phenotype.id\"=\"id.genotype\"))\n",
    "    id.genotype.match <- phenotype.id.merge$index\n",
    "    \n",
    "    ## Genotype\n",
    "    Geno <- seqGetData(genofile, \"$dosage\")\n",
    "    Geno <- Geno[id.genotype.match,,drop=FALSE]\n",
    "    \n",
    "    summary_stat <- NULL\n",
    "    GTSinvG_rare <- NULL\n",
    "    cov_cond <- NULL\n",
    "    \n",
    "    if(!is.null(Geno))\n",
    "    {\n",
    "      # Summary statistics\n",
    "      genotype <- matrix_impute(Geno)\n",
    "      variant_info <- data.frame(chr,pos,ref,alt,row.names=NULL)\n",
    "      \n",
    "      ## Annotation\n",
    "      Anno.Int.PHRED.sub.category <- Anno.Int.PHRED.sub[lof.in.dmissense,]\n",
    "      \n",
    "      try(summary_stat <- MetaSTAARlite_worker_sumstat(genotype,obj_nullmodel,variant_info,qc_label,\n",
    "                                                       Anno.Int.PHRED.sub.category),silent=silent)\n",
    "      \n",
    "      # Covariance matrices\n",
    "      genotype <- matrix_flip(Geno)$Geno\n",
    "      genotype <- as(genotype,\"dgCMatrix\")\n",
    "      \n",
    "      rm(Geno)\n",
    "      gc()\n",
    "      \n",
    "      try(GTSinvG_rare <- MetaSTAARlite_worker_cov(genotype,obj_nullmodel,cov_maf_cutoff,\n",
    "                                                   qc_label,signif.digits),silent=silent)\n",
    "      \n",
    "      # Covariance matrices for conditional analysis\n",
    "      if(!is.null(known_loci))\n",
    "      {\n",
    "        try(cov_cond <- MetaSTAARlite_worker_cov_cond(genotype,G_SNV,obj_nullmodel,variant_info,variant_adj_info),silent=silent)\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    summary_stat_list[[\"disruptive_missense\"]] <- summary_stat\n",
    "    GTSinvG_rare_list[[\"disruptive_missense\"]] <- GTSinvG_rare\n",
    "    cov_cond_list[[\"disruptive_missense\"]] <- cov_cond\n",
    "    \n",
    "    seqResetFilter(genofile)\n",
    "    \n",
    "    if(!is.null(known_loci))\n",
    "    {\n",
    "      return(list(summary_stat_list=summary_stat_list,\n",
    "                  GTSinvG_rare_list=GTSinvG_rare_list,\n",
    "                  cov_cond_list=cov_cond_list))\n",
    "    }else\n",
    "    {\n",
    "      return(list(summary_stat_list=summary_stat_list,\n",
    "                  GTSinvG_rare_list=GTSinvG_rare_list))\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  \n",
    "  ###########################################################\n",
    "  #           User Input\n",
    "  ###########################################################\n",
    "  ## Null model\n",
    "  obj_nullmodel <- get(load(Null_Model))\n",
    "\n",
    "## Parameter\n",
    "QC_label <- \"annotation/filter\"\n",
    "geno_missing_imputation <- \"mean\"\n",
    "variant_type <- \"SNV\"\n",
    "\n",
    "## Annotation_dir\n",
    "Annotation_dir <- \"annotation/info/FunctionalAnnotation\"\n",
    "## Annotation channel\n",
    "Annotation_name_catalog <- read.csv(Annotation_name_catalog)\n",
    "## Use_annotation_weights\n",
    "Use_annotation_weights <- TRUE\n",
    "## Annotation name\n",
    "Annotation_name <- c(\"CADD\",\"LINSIGHT\",\"FATHMM.XF\",\"aPC.EpigeneticActive\",\"aPC.EpigeneticRepressed\",\"aPC.EpigeneticTranscription\",\n",
    "                     \"aPC.Conservation\",\"aPC.LocalDiversity\",\"aPC.Mappability\",\"aPC.TF\",\"aPC.Protein\")\n",
    "\n",
    "agds_dir <- vector()\n",
    "agds_dir[2] <- GDS_File_Chr2\n",
    "agds_dir[5] <- GDS_File_Chr5\n",
    "    \n",
    "  ## output file name\n",
    "  output_file_name <- paste0(OUTPUT_PATH,\"/AoU_TC_Coding\")\n",
    "  \n",
    "  ###########################################################\n",
    "  #           Main Function \n",
    "  ###########################################################\n",
    "  ## gene number in job\n",
    "  gene_num_in_array <- 50\n",
    "  group.num.allchr <- ceiling(table(genes_info[,2])/gene_num_in_array)\n",
    "  sum(group.num.allchr)\n",
    "  \n",
    "  ## analyze large coding masks\n",
    "  arrayid <- c(57,112,112,113,113,113,113,113,113,113)\n",
    "  sub_seq_id <- c(840,543,544,575,576,577,578,579,580,582)\n",
    "  \n",
    "  region_spec <- data.frame(arrayid,sub_seq_id) \n",
    "  sub_seq_id <- ((arrayid_longmask-1)*5+1):min(arrayid_longmask*5,length(arrayid))\n",
    "  \n",
    "  genes <- genes_info\n",
    "  \n",
    "  coding_sumstat <- list()\n",
    "  coding_cov <- list()\n",
    "  for(kk in sub_seq_id){\n",
    "    print(kk)\n",
    "    arrayid <- region_spec$arrayid[kk]\n",
    "    sub_id <- region_spec$sub_seq_id[kk]\n",
    "    \n",
    "    chr <- which.max(arrayid <= cumsum(group.num.allchr))\n",
    "    \n",
    "    ## aGDS file\n",
    "    agds.path <- agds_dir[chr]\n",
    "    genofile <- seqOpen(agds.path)\n",
    "    \n",
    "    genes_info_chr <- genes_info[genes_info[,2]==chr,]\n",
    "    gene_name <- genes_info_chr[sub_id,1]\n",
    "    results_temp <- coding_MetaSTAARlite_worker(chr=chr,gene_name=gene_name,genofile=genofile,obj_nullmodel=obj_nullmodel,genes=genes,\n",
    "                                                cov_maf_cutoff=0.05,signif.digits=NULL,\n",
    "                                                QC_label=QC_label,check_qc_label=TRUE,variant_type=variant_type,\n",
    "                                                Annotation_dir=Annotation_dir,Annotation_name_catalog=Annotation_name_catalog,\n",
    "                                                Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name)\n",
    "    coding_sumstat[[gene_name]] <- results_temp$summary_stat_list\n",
    "    coding_cov[[gene_name]] <- results_temp$GTSinvG_rare_list\n",
    "    \n",
    "    seqClose(genofile)\n",
    "  }\n",
    "  \n",
    "  save(coding_sumstat,file=paste0(output_file_name,\"_sumstat_\",arrayid_longmask+379,\".Rdata\"),compress = \"xz\")\n",
    "  save(coding_cov,file=paste0(output_file_name,\"_cov_\",arrayid_longmask+379,\".Rdata\"),compress = \"xz\")\n",
    "  \n",
    "})[3]\n",
    "\n",
    "save(time,file = paste0(output_file_name,\"_time_\",arrayid_longmask+379,\".Rdata\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Meta_Coding_AoU_LongMasks.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile Meta_Coding_AoU_LongMasks.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${arrayid} ${Null_Model} ${Annotation_name_catalog} ${GDS_File_Chr2} ${GDS_File_Chr5} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for(arrayid in c(1,2)){\n",
    "        tasks <- rbind(tasks, data.frame(\n",
    "            '--env arrayid'=arrayid,\n",
    "            '--input Null_Model'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/MetaSTAAR/TC_Null_Model.RData\",\n",
    "            '--input Annotation_name_catalog'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataGDS/acaf_threshold_v7/Annotation_name_catalog.csv\",\n",
    "            '--input GDS_File_Chr2'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataAGDS/exome_v7.1/exome.chr\",2,\".gds\"),\n",
    "            '--input GDS_File_Chr5'=paste0(\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/dataAGDS/exome_v7.1/exome.chr\",5,\".gds\"),\n",
    "            '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/MetaSTAAR/Meta_Coding_AoU_LongMasks.R\",\n",
    "            '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/MetaSTAAR/AoU_Sumstats/\",\n",
    "            check.names = FALSE\n",
    "        )) \n",
    "    }\n",
    "  \n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://Meta_Coding_AoU_LongMasks.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][ 28.3 KiB/ 28.3 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/28.3 KiB.                                     \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp Meta_Coding_AoU_LongMasks.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/MetaSTAAR/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: meta-codin--williamsjacr--241118-144831-35\n",
      "  job-name: meta-coding-aou-longmasks\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15715809343684227583\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/14475785995333795255\n",
      "Launched job-id: meta-codin--williamsjacr--241118-144831-35\n",
      "2 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'meta-codin--williamsjacr--241118-144831-35' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'meta-codin--williamsjacr--241118-144831-35' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 128 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script Meta_Coding_AoU_LongMasks.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status    Last Update\r\n",
      "---------------  ------  --------  --------------\r\n",
      "meta-coding-...       2  Success   11-18 15:15:40\r\n",
      "meta-coding-...       1  Success   11-18 15:33:54\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'meta-codin--williamsjacr--241118-144831-35' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "code_folding": [
     23,
     180
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing Meta_Coding_UKB_AoU.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile Meta_Coding_UKB_AoU.R\n",
    "rm(list=ls())\n",
    "gc()\n",
    "\n",
    "arrayid <- as.numeric(commandArgs(TRUE)[1])\n",
    "\n",
    "UKB_Sumstats <- commandArgs(TRUE)[2]\n",
    "print(UKB_Sumstats)\n",
    "\n",
    "AoU_Sumstats <- commandArgs(TRUE)[3]\n",
    "print(AoU_Sumstats)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[4]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "time <- system.time({\n",
    "    ## load required packages\n",
    "library(STAAR)\n",
    "library(STAARpipeline)\n",
    "library(MetaSTAAR)\n",
    "library(dplyr)\n",
    "library(Matrix)\n",
    "\n",
    "MetaSTAARlite_merge <- function(chr,sample.sizes,sumstat.list,cov.list,\n",
    "                                rare_maf_cutoff=0.01,cov_maf_cutoff,\n",
    "                                check_qc_label=FALSE,variant_type=c(\"SNV\",\"Indel\",\"variant\"),\n",
    "                                Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL){\n",
    "  \n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "  \n",
    "  cov_maf_cutoff[cov_maf_cutoff == 0.5] <- 0.5 + 1e-16\n",
    "  ### summary statistics\n",
    "  sumstat.list <- lapply(sumstat.list, function(x) {\n",
    "    if (!(is.null(x)) && !(\"qc_label\" %in% colnames(x))){\n",
    "      data.frame(x[,1:4],qc_label=\"PASS\",x[,5:dim(x)[2]],stringsAsFactors = FALSE)\n",
    "    }else{\n",
    "      x\n",
    "    }\n",
    "  })\n",
    "  sumstat.varid.list <- mapply(function(x,y) {\n",
    "    x[(x$MAF<y)&(x$MAF>0)&(x$qc_label==\"PASS\"),1:4]\n",
    "  }, x = sumstat.list, y = cov_maf_cutoff, SIMPLIFY = FALSE)\n",
    "  sumstat.varid.merge <- do.call(\"rbind\",sumstat.varid.list)\n",
    "  sumstat.varid.nodup <- sumstat.varid.merge[!duplicated(sumstat.varid.merge),]\n",
    "  if (is.null(sumstat.varid.nodup)) {\n",
    "    return(NULL)\n",
    "  }else if (dim(sumstat.varid.nodup)[1] == 0) {\n",
    "    return(NULL)\n",
    "  }\n",
    "  sumstat.merge.list <- lapply(sumstat.list, function(x) {\n",
    "    if (is.null(x)) {\n",
    "      cbind(sumstat.varid.nodup,qc_label=NA,alt_AC=NA,MAC=NA,MAF=NA,N=NA,U=NA,V=NA,\"1\"=NA)\n",
    "    }else {\n",
    "      left_join(sumstat.varid.nodup,x,by=c(\"chr\"=\"chr\",\n",
    "                                           \"pos\"=\"pos\",\n",
    "                                           \"ref\"=\"ref\",\n",
    "                                           \"alt\"=\"alt\"))\n",
    "    }\n",
    "  })\n",
    "  sumstat.merge.list <- mapply(function(x,y) {\n",
    "    x[is.na(x[,\"N\"]),\"N\"] <- y\n",
    "    x[is.na(x[,\"qc_label\"]),\"qc_label\"] <- \"PASS\"\n",
    "    x[is.na(x)] <- 0\n",
    "    return(x)\n",
    "  }, x = sumstat.merge.list, y = sample.sizes, SIMPLIFY = FALSE)\n",
    "  sumstat.varid.nodup$index <- 1:dim(sumstat.varid.nodup)[1]\n",
    "  sumstat.index.list <- lapply(sumstat.varid.list, function(x) {\n",
    "    if (is.null(x)) {\n",
    "      integer(0)\n",
    "    }else{\n",
    "      left_join(x,sumstat.varid.nodup,by=c(\"chr\"=\"chr\",\n",
    "                                           \"pos\"=\"pos\",\n",
    "                                           \"ref\"=\"ref\",\n",
    "                                           \"alt\"=\"alt\"))$index\n",
    "    }\n",
    "  })\n",
    "  rm(sumstat.list,sumstat.varid.list,sumstat.varid.merge)\n",
    "  gc()\n",
    "  \n",
    "  ### covariance matrices\n",
    "  cov.list <- lapply(cov.list, function(x) {\n",
    "    if (is.null(x)) {\n",
    "      as(matrix(nrow=0,ncol=0),\"dgCMatrix\")\n",
    "    }else if (dim(x)[1] == 0) {\n",
    "      as(matrix(nrow=0,ncol=0),\"dgCMatrix\")\n",
    "    }else {\n",
    "      x[,1:dim(x)[1]]\n",
    "    }\n",
    "  })\n",
    "  \n",
    "  cov.null <- matrix(0, nrow = dim(sumstat.varid.nodup)[1], ncol = dim(sumstat.varid.nodup)[1])\n",
    "  cov.merge.list <- mapply(function(x,y) {\n",
    "    if (length(x) == 1) {\n",
    "      # in this case y is a scalar\n",
    "      cov.null[x,x] <- y\n",
    "    }else if (length(x) > 1) {\n",
    "      cov.null[x,x] <- as.matrix(forceSymmetric(y, uplo = \"U\"))\n",
    "    }\n",
    "    return(cov.null)\n",
    "  }, x = sumstat.index.list, y = cov.list, SIMPLIFY = FALSE)\n",
    "  \n",
    "  ### select rare variant based on the input cutoff\n",
    "  alt_AC.merge <- as.integer(Reduce(\"+\",lapply(sumstat.merge.list, function(x) {x$alt_AC})))\n",
    "  N.merge.nonzero <- Reduce(\"+\",lapply(sumstat.merge.list, function(x) {x$N * (x$MAC != 0)}))\n",
    "  alt_AF.merge.nonzero <- alt_AC.merge / (2 * N.merge.nonzero)\n",
    "  N.merge.zero <- Reduce(\"+\",lapply(sumstat.merge.list, function(x) {x$N * (x$MAC == 0)}))\n",
    "  alt_AC.merge <- alt_AC.merge + (alt_AF.merge.nonzero > 0.5) * (2 * N.merge.zero)\n",
    "  N.merge <- Reduce(\"+\",lapply(sumstat.merge.list, function(x) {x$N}))\n",
    "  MAC.merge <- pmin(alt_AC.merge, 2 * N.merge - alt_AC.merge)\n",
    "  MAF.merge <- MAC.merge / (2 * N.merge)\n",
    "  rv.index <- (MAF.merge<rare_maf_cutoff) & Reduce(\"*\",mapply(function(x,y) {\n",
    "    x$MAF<y\n",
    "  }, x = sumstat.merge.list, y = cov_maf_cutoff, SIMPLIFY = FALSE))\n",
    "  if (check_qc_label){\n",
    "    rv.index <- rv.index & Reduce(\"*\",lapply(sumstat.merge.list, function(x) {\n",
    "      x$qc_label==\"PASS\"\n",
    "    }))\n",
    "  }\n",
    "  \n",
    "  info <- cbind(sumstat.varid.nodup[,c(\"chr\",\"pos\",\"ref\",\"alt\")],\n",
    "                MAC=MAC.merge,MAF=MAF.merge)[rv.index,]\n",
    "  \n",
    "  U.merge <- Reduce(\"+\", lapply(sumstat.merge.list, function(x) {\n",
    "    x <- x[rv.index,]\n",
    "    return(x$U)\n",
    "  }))\n",
    "  \n",
    "  cov.merge <- Reduce(\"+\", mapply(function(x,y) {\n",
    "    (x - as.matrix(y[,(which(colnames(y)==\"V\")+1):dim(y)[2]]) %*% t(as.matrix(y[,(which(colnames(y)==\"V\")+1):dim(y)[2]])))[rv.index, rv.index]\n",
    "  }, x = cov.merge.list, y = sumstat.merge.list, SIMPLIFY = FALSE))\n",
    "  \n",
    "  ## Annotation\n",
    "  Anno.Int.PHRED.sub <- NULL\n",
    "  Anno.Int.PHRED.sub.name <- NULL\n",
    "  \n",
    "  if(variant_type==\"SNV\")\n",
    "  {\n",
    "    if(Use_annotation_weights)\n",
    "    {\n",
    "      for(k in 1:length(Annotation_name))\n",
    "      {\n",
    "        Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,Annotation_name[k])\n",
    "        Annotation.list <- lapply(sumstat.merge.list, function(x) {\n",
    "          if(Annotation_name[k]%in%colnames(x)) {\n",
    "            x[[Annotation_name[k]]]\n",
    "          }else {\n",
    "            0\n",
    "          }\n",
    "        })\n",
    "        Annotation.PHRED.sum <- Reduce(\"+\", Annotation.list)\n",
    "        Annotation.PHRED.gt0 <- Reduce(\"+\",lapply(Annotation.list, function(x) {x > 0}))\n",
    "        Annotation.PHRED <- Annotation.PHRED.sum / Annotation.PHRED.gt0\n",
    "        Annotation.PHRED[!is.finite(Annotation.PHRED)] <- 0\n",
    "        \n",
    "        if(Annotation_name[k]==\"aPC.LocalDiversity\")\n",
    "        {\n",
    "          Annotation.PHRED.2 <- -10*log10(1-10^(-Annotation.PHRED/10))\n",
    "          Annotation.PHRED <- cbind(Annotation.PHRED,Annotation.PHRED.2)\n",
    "          Anno.Int.PHRED.sub.name <- c(Anno.Int.PHRED.sub.name,paste0(Annotation_name[k],\"(-)\"))\n",
    "        }\n",
    "        Anno.Int.PHRED.sub <- cbind(Anno.Int.PHRED.sub,Annotation.PHRED)\n",
    "      }\n",
    "      \n",
    "      Anno.Int.PHRED.sub <- data.frame(Anno.Int.PHRED.sub)\n",
    "      colnames(Anno.Int.PHRED.sub) <- c(Anno.Int.PHRED.sub.name)\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  annotation_phred <- Anno.Int.PHRED.sub[rv.index,]\n",
    "  \n",
    "  rm(list=setdiff(ls(), c(\"info\",\"U.merge\",\"cov.merge\",\"annotation_phred\")))\n",
    "  gc()\n",
    "  \n",
    "  return(list(info=info,\n",
    "              U=U.merge,\n",
    "              cov=as.matrix(cov.merge),\n",
    "              annotation_phred=annotation_phred))\n",
    "}\n",
    "\n",
    "coding_MetaSTAARlite <- function(chr,gene_name,genes,\n",
    "                                 sample.sizes,coding_sumstat_gene_list,coding_cov_gene_list,\n",
    "                                 cov_maf_cutoff,rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                                 check_qc_label=FALSE,variant_type=c(\"SNV\",\"Indel\",\"variant\"),\n",
    "                                 Use_annotation_weights=c(TRUE,FALSE),Annotation_name=NULL,silent=FALSE){\n",
    "\n",
    "  ## evaluate choices\n",
    "  variant_type <- match.arg(variant_type)\n",
    "\n",
    "  ### Gene\n",
    "  kk <- which(genes[,1]==gene_name)\n",
    "\n",
    "  ################################################\n",
    "  #                  plof_ds\n",
    "  ################################################\n",
    "  results_plof_ds <- c()\n",
    "\n",
    "  sumstat.list <- lapply(coding_sumstat_gene_list, function(x) {\n",
    "    x[[\"plof_ds\"]]\n",
    "  })\n",
    "  cov.list <- lapply(coding_cov_gene_list, function(x) {\n",
    "    x[[\"plof_ds\"]]\n",
    "  })\n",
    "\n",
    "  gene_test_merge <- MetaSTAARlite_merge(chr,sample.sizes,sumstat.list,cov.list,\n",
    "                                         cov_maf_cutoff=cov_maf_cutoff,rare_maf_cutoff=rare_maf_cutoff,\n",
    "                                         check_qc_label=check_qc_label,variant_type=variant_type,\n",
    "                                         Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name)\n",
    "  if(length(gene_test_merge$U)>=2)\n",
    "  {\n",
    "    ## Annotation\n",
    "    annotation_phred <- gene_test_merge$annotation_phred\n",
    "    pvalues <- 0\n",
    "    try(pvalues <- MetaSTAAR(gene_test_merge,annotation_phred,rv_num_cutoff),silent=silent)\n",
    "\n",
    "    if(inherits(pvalues,\"list\"))\n",
    "    {\n",
    "      results_temp <- as.vector(genes[kk,])\n",
    "      results_temp[3] <- \"plof_ds\"\n",
    "      results_temp[2] <- chr\n",
    "      results_temp[1] <- as.character(genes[kk,1])\n",
    "      results_temp[4] <- pvalues$num_variant\n",
    "\n",
    "\n",
    "      results_temp <- c(results_temp,pvalues$results_MetaSTAAR_S_1_25,pvalues$results_MetaSTAAR_S_1_1,\n",
    "                        pvalues$results_MetaSTAAR_B_1_25,pvalues$results_MetaSTAAR_B_1_1,pvalues$results_MetaSTAAR_A_1_25,\n",
    "                        pvalues$results_MetaSTAAR_A_1_1,pvalues$results_ACAT_O_MS,pvalues$results_MetaSTAAR_O)\n",
    "\n",
    "      results_plof_ds <- rbind(results_plof_ds,results_temp)\n",
    "    }\n",
    "  }\n",
    "\n",
    "  if(!is.null(results_plof_ds))\n",
    "  {\n",
    "    colnames(results_plof_ds) <- colnames(results_plof_ds, do.NULL = FALSE, prefix = \"col\")\n",
    "    colnames(results_plof_ds)[1:4] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\")\n",
    "    colnames(results_plof_ds)[(dim(results_plof_ds)[2]-1):dim(results_plof_ds)[2]] <- c(\"ACAT-O-MS\",\"MetaSTAAR-O\")\n",
    "  }\n",
    "\n",
    "  #####################################################\n",
    "  #                      plof\n",
    "  #####################################################\n",
    "  results_plof <- c()\n",
    "\n",
    "  sumstat.list <- lapply(coding_sumstat_gene_list, function(x) {\n",
    "    x[[\"plof\"]]\n",
    "  })\n",
    "  cov.list <- lapply(coding_cov_gene_list, function(x) {\n",
    "    x[[\"plof\"]]\n",
    "  })\n",
    "\n",
    "  gene_test_merge <- MetaSTAARlite_merge(chr,sample.sizes,sumstat.list,cov.list,\n",
    "                                         cov_maf_cutoff=cov_maf_cutoff,rare_maf_cutoff=rare_maf_cutoff,\n",
    "                                         check_qc_label=check_qc_label,variant_type=variant_type,\n",
    "                                         Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name)\n",
    "  if(length(gene_test_merge$U)>=2)\n",
    "  {\n",
    "    ## Annotation\n",
    "    annotation_phred <- gene_test_merge$annotation_phred\n",
    "    pvalues <- 0\n",
    "    try(pvalues <- MetaSTAAR(gene_test_merge,annotation_phred,rv_num_cutoff),silent=silent)\n",
    "\n",
    "    if(inherits(pvalues,\"list\"))\n",
    "    {\n",
    "      results_temp <- as.vector(genes[kk,])\n",
    "      results_temp[3] <- \"plof\"\n",
    "      results_temp[2] <- chr\n",
    "      results_temp[1] <- as.character(genes[kk,1])\n",
    "      results_temp[4] <- pvalues$num_variant\n",
    "\n",
    "\n",
    "      results_temp <- c(results_temp,pvalues$results_MetaSTAAR_S_1_25,pvalues$results_MetaSTAAR_S_1_1,\n",
    "                        pvalues$results_MetaSTAAR_B_1_25,pvalues$results_MetaSTAAR_B_1_1,pvalues$results_MetaSTAAR_A_1_25,\n",
    "                        pvalues$results_MetaSTAAR_A_1_1,pvalues$results_ACAT_O_MS,pvalues$results_MetaSTAAR_O)\n",
    "\n",
    "      results_plof <- rbind(results_plof,results_temp)\n",
    "    }\n",
    "  }\n",
    "\n",
    "  if(!is.null(results_plof))\n",
    "  {\n",
    "    colnames(results_plof) <- colnames(results_plof, do.NULL = FALSE, prefix = \"col\")\n",
    "    colnames(results_plof)[1:4] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\")\n",
    "    colnames(results_plof)[(dim(results_plof)[2]-1):dim(results_plof)[2]] <- c(\"ACAT-O-MS\",\"MetaSTAAR-O\")\n",
    "  }\n",
    "\n",
    "  #############################################\n",
    "  #             synonymous\n",
    "  #############################################\n",
    "  results_synonymous <- c()\n",
    "\n",
    "  sumstat.list <- lapply(coding_sumstat_gene_list, function(x) {\n",
    "    x[[\"synonymous\"]]\n",
    "  })\n",
    "  cov.list <- lapply(coding_cov_gene_list, function(x) {\n",
    "    x[[\"synonymous\"]]\n",
    "  })\n",
    "\n",
    "  gene_test_merge <- MetaSTAARlite_merge(chr,sample.sizes,sumstat.list,cov.list,\n",
    "                                         cov_maf_cutoff=cov_maf_cutoff,rare_maf_cutoff=rare_maf_cutoff,\n",
    "                                         check_qc_label=check_qc_label,variant_type=variant_type,\n",
    "                                         Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name)\n",
    "  if(length(gene_test_merge$U)>=2)\n",
    "  {\n",
    "    ## Annotation\n",
    "    annotation_phred <- gene_test_merge$annotation_phred\n",
    "    pvalues <- 0\n",
    "    try(pvalues <- MetaSTAAR(gene_test_merge,annotation_phred,rv_num_cutoff),silent=silent)\n",
    "\n",
    "    if(inherits(pvalues,\"list\"))\n",
    "    {\n",
    "      results_temp <- as.vector(genes[kk,])\n",
    "      results_temp[3] <- \"synonymous\"\n",
    "      results_temp[2] <- chr\n",
    "      results_temp[1] <- as.character(genes[kk,1])\n",
    "      results_temp[4] <- pvalues$num_variant\n",
    "\n",
    "\n",
    "      results_temp <- c(results_temp,pvalues$results_MetaSTAAR_S_1_25,pvalues$results_MetaSTAAR_S_1_1,\n",
    "                        pvalues$results_MetaSTAAR_B_1_25,pvalues$results_MetaSTAAR_B_1_1,pvalues$results_MetaSTAAR_A_1_25,\n",
    "                        pvalues$results_MetaSTAAR_A_1_1,pvalues$results_ACAT_O_MS,pvalues$results_MetaSTAAR_O)\n",
    "\n",
    "      results_synonymous <- rbind(results_synonymous,results_temp)\n",
    "    }\n",
    "  }\n",
    "\n",
    "  if(!is.null(results_synonymous))\n",
    "  {\n",
    "    colnames(results_synonymous) <- colnames(results_synonymous, do.NULL = FALSE, prefix = \"col\")\n",
    "    colnames(results_synonymous)[1:4] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\")\n",
    "    colnames(results_synonymous)[(dim(results_synonymous)[2]-1):dim(results_synonymous)[2]] <- c(\"ACAT-O-MS\",\"MetaSTAAR-O\")\n",
    "  }\n",
    "\n",
    "  #################################################\n",
    "  #        missense\n",
    "  #################################################\n",
    "  results <- c()\n",
    "\n",
    "  sumstat.list <- lapply(coding_sumstat_gene_list, function(x) {\n",
    "    x[[\"missense\"]]\n",
    "  })\n",
    "  cov.list <- lapply(coding_cov_gene_list, function(x) {\n",
    "    x[[\"missense\"]]\n",
    "  })\n",
    "\n",
    "  gene_test_merge <- MetaSTAARlite_merge(chr,sample.sizes,sumstat.list,cov.list,\n",
    "                                         cov_maf_cutoff=cov_maf_cutoff,rare_maf_cutoff=rare_maf_cutoff,\n",
    "                                         check_qc_label=check_qc_label,variant_type=variant_type,\n",
    "                                         Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name)\n",
    "  if(length(gene_test_merge$U)>=2)\n",
    "  {\n",
    "    ## Annotation\n",
    "    annotation_phred <- gene_test_merge$annotation_phred\n",
    "    pvalues <- 0\n",
    "    try(pvalues <- MetaSTAAR(gene_test_merge,annotation_phred,rv_num_cutoff),silent=silent)\n",
    "\n",
    "    if(inherits(pvalues,\"list\"))\n",
    "    {\n",
    "      results_temp <- as.vector(genes[kk,])\n",
    "      results_temp[3] <- \"missense\"\n",
    "      results_temp[2] <- chr\n",
    "      results_temp[1] <- as.character(genes[kk,1])\n",
    "      results_temp[4] <- pvalues$num_variant\n",
    "\n",
    "\n",
    "      results_temp <- c(results_temp,pvalues$results_MetaSTAAR_S_1_25,pvalues$results_MetaSTAAR_S_1_1,\n",
    "                        pvalues$results_MetaSTAAR_B_1_25,pvalues$results_MetaSTAAR_B_1_1,pvalues$results_MetaSTAAR_A_1_25,\n",
    "                        pvalues$results_MetaSTAAR_A_1_1,pvalues$results_ACAT_O_MS,pvalues$results_MetaSTAAR_O)\n",
    "\n",
    "      results <- rbind(results,results_temp)\n",
    "    }\n",
    "  }\n",
    "\n",
    "  #################################################\n",
    "  #         disruptive missense\n",
    "  #################################################\n",
    "  sumstat.list <- lapply(coding_sumstat_gene_list, function(x) {\n",
    "    x[[\"disruptive_missense\"]]\n",
    "  })\n",
    "  cov.list <- lapply(coding_cov_gene_list, function(x) {\n",
    "    x[[\"disruptive_missense\"]]\n",
    "  })\n",
    "\n",
    "  gene_test_merge <- MetaSTAARlite_merge(chr,sample.sizes,sumstat.list,cov.list,\n",
    "                                         cov_maf_cutoff=cov_maf_cutoff,rare_maf_cutoff=rare_maf_cutoff,\n",
    "                                         check_qc_label=check_qc_label,variant_type=variant_type,\n",
    "                                         Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name)\n",
    "  if(length(gene_test_merge$U)>=2)\n",
    "  {\n",
    "    ## Annotation\n",
    "    annotation_phred <- gene_test_merge$annotation_phred\n",
    "    pvalues <- 0\n",
    "    try(pvalues <- MetaSTAAR(gene_test_merge,annotation_phred,rv_num_cutoff),silent=silent)\n",
    "\n",
    "    if(inherits(pvalues,\"list\"))\n",
    "    {\n",
    "      results_temp <- as.vector(genes[kk,])\n",
    "      results_temp[3] <- \"disruptive_missense\"\n",
    "      results_temp[2] <- chr\n",
    "      results_temp[1] <- as.character(genes[kk,1])\n",
    "      results_temp[4] <- pvalues$num_variant\n",
    "\n",
    "\n",
    "      results_temp <- c(results_temp,pvalues$results_MetaSTAAR_S_1_25,pvalues$results_MetaSTAAR_S_1_1,\n",
    "                        pvalues$results_MetaSTAAR_B_1_25,pvalues$results_MetaSTAAR_B_1_1,pvalues$results_MetaSTAAR_A_1_25,\n",
    "                        pvalues$results_MetaSTAAR_A_1_1,pvalues$results_ACAT_O_MS,pvalues$results_MetaSTAAR_O)\n",
    "\n",
    "      results <- rbind(results,results_temp)\n",
    "    }\n",
    "  }\n",
    "\n",
    "  if(!is.null(results))\n",
    "  {\n",
    "    colnames(results) <- colnames(results, do.NULL = FALSE, prefix = \"col\")\n",
    "    colnames(results)[1:4] <- c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\")\n",
    "    colnames(results)[(dim(results)[2]-1):dim(results)[2]] <- c(\"ACAT-O-MS\",\"MetaSTAAR-O\")\n",
    "\n",
    "    if(dim(results)[1]==1)\n",
    "    {\n",
    "      if(results[3]!=\"disruptive_missense\")\n",
    "      {\n",
    "        results <- cbind(results,matrix(1,1,6))\n",
    "        colnames(results)[(dim(results)[2]-5):dim(results)[2]] <- c(\"SKAT-MS(1,25)-Disruptive\",\"SKAT-MS(1,1)-Disruptive\",\"Burden-MS(1,25)-Disruptive\",\"Burden-MS(1,1)-Disruptive\",\"ACAT-V-MS(1,25)-Disruptive\",\"ACAT-V-MS(1,1)-Disruptive\")\n",
    "        results_missense <- results\n",
    "        results_ds <- c()\n",
    "      }else\n",
    "      {\n",
    "        results_missense <- c()\n",
    "        results_ds <- results\n",
    "        results <- c()\n",
    "      }\n",
    "    }\n",
    "\n",
    "    if(!is.null(results))\n",
    "    {\n",
    "      if(dim(results)[1]==2)\n",
    "      {\n",
    "        results_m <- c(results[1,],rep(0,6))\n",
    "        names(results_m)[(length(results_m)-5):length(results_m)] <- c(\"SKAT-MS(1,25)-Disruptive\",\"SKAT-MS(1,1)-Disruptive\",\"Burden-MS(1,25)-Disruptive\",\"Burden-MS(1,1)-Disruptive\",\"ACAT-V-MS(1,25)-Disruptive\",\"ACAT-V-MS(1,1)-Disruptive\")\n",
    "        results_m[(length(results_m)-5):length(results_m)] <- results[2,c(\"SKAT-MS(1,25)\",\"SKAT-MS(1,1)\",\"Burden-MS(1,25)\",\"Burden-MS(1,1)\",\"ACAT-V-MS(1,25)\",\"ACAT-V-MS(1,1)\")]\n",
    "        apc_num <- (length(results_m)-18)/6\n",
    "        p_seq <- c(1:apc_num,1:apc_num+(apc_num+1),1:apc_num+2*(apc_num+1),1:apc_num+3*(apc_num+1),1:apc_num+4*(apc_num+1),1:apc_num+5*(apc_num+1),(6*apc_num+9):(6*apc_num+14))\n",
    "        results_m[\"MetaSTAAR-O\"] <- CCT(as.numeric(results_m[5:length(results_m)][p_seq]))\n",
    "        results_m[\"MetaSTAAR-S(1,25)\"] <- CCT(as.numeric(results_m[5:length(results_m)][c(1:apc_num,6*apc_num+9)]))\n",
    "        results_m[\"MetaSTAAR-S(1,1)\"] <- CCT(as.numeric(results_m[5:length(results_m)][c(1:apc_num+(apc_num+1),6*apc_num+10)]))\n",
    "        results_m[\"MetaSTAAR-B(1,25)\"] <- CCT(as.numeric(results_m[5:length(results_m)][c(1:apc_num+2*(apc_num+1),6*apc_num+11)]))\n",
    "        results_m[\"MetaSTAAR-B(1,1)\"] <- CCT(as.numeric(results_m[5:length(results_m)][c(1:apc_num+3*(apc_num+1),6*apc_num+12)]))\n",
    "        results_m[\"MetaSTAAR-A(1,25)\"] <- CCT(as.numeric(results_m[5:length(results_m)][c(1:apc_num+4*(apc_num+1),6*apc_num+13)]))\n",
    "        results_m[\"MetaSTAAR-A(1,1)\"] <- CCT(as.numeric(results_m[5:length(results_m)][c(1:apc_num+5*(apc_num+1),6*apc_num+14)]))\n",
    "\n",
    "        results_ds <- c()\n",
    "        results_ds <- rbind(results_ds,results[2,])\n",
    "\n",
    "        results <- c()\n",
    "        results <- rbind(results,results_m)\n",
    "      }\n",
    "    }\n",
    "  }else\n",
    "  {\n",
    "    results <- c()\n",
    "    results_ds <- c()\n",
    "  }\n",
    "\n",
    "  results_coding <- list(plof=results_plof,plof_ds=results_plof_ds,missense=results,disruptive_missense=results_ds,synonymous=results_synonymous)\n",
    "\n",
    "  return(results_coding)\n",
    "}\n",
    "\n",
    "###########################################################\n",
    "#           User Input\n",
    "###########################################################\n",
    "## Directories of the study-specific summary statistics file folders\n",
    "file.dir <- c(paste0(UKB_Sumstats,\"/\"),\n",
    "              paste0(AoU_Sumstats,\"/\"))\n",
    "file.prefix <- c(\"UKB_TC_Coding\",\"AoU_TC_Coding\")\n",
    "## Sample sizes of participating studies\n",
    "sample.sizes <- c(446933,94532)\n",
    "\n",
    "## variant_type\n",
    "variant_type <- \"SNV\"\n",
    "## cov_maf_cutoff\n",
    "cov_maf_cutoff <- c(0.05,0.05)\n",
    "\n",
    "## Use_annotation_weights\n",
    "Use_annotation_weights <- TRUE\n",
    "## Annotation name\n",
    "Annotation_name <- c(\"CADD\",\"LINSIGHT\",\"FATHMM.XF\",\"aPC.EpigeneticActive\",\"aPC.EpigeneticRepressed\",\"aPC.EpigeneticTranscription\",\n",
    "                     \"aPC.Conservation\",\"aPC.LocalDiversity\",\"aPC.Mappability\",\"aPC.TF\",\"aPC.Protein\")\n",
    "\n",
    "## output path\n",
    "output_path <- OUTPUT_PATH\n",
    "## output file name\n",
    "output_file_name <- \"UKB_AoU_Coding\"\n",
    "\n",
    "###########################################################\n",
    "#           Main Function \n",
    "###########################################################\n",
    "## gene number in job\n",
    "gene_num_in_array <- 50\n",
    "group.num.allchr <- ceiling(table(genes_info[,2])/gene_num_in_array)\n",
    "sum(group.num.allchr)\n",
    "\n",
    "chr <- as.integer(which.max(arrayid <= cumsum(group.num.allchr)))\n",
    "group.num <- group.num.allchr[chr]\n",
    "\n",
    "if (chr == 1){\n",
    "  groupid <- arrayid\n",
    "}else{\n",
    "  groupid <- arrayid - cumsum(group.num.allchr)[chr-1]\n",
    "}\n",
    "\n",
    "genes_info_chr <- genes_info[genes_info[,2]==chr,]\n",
    "sub_seq_num <- dim(genes_info_chr)[1]\n",
    "\n",
    "if(groupid < group.num)\n",
    "{\n",
    "  sub_seq_id <- ((groupid - 1)*gene_num_in_array + 1):(groupid*gene_num_in_array)\n",
    "}else\n",
    "{\n",
    "  sub_seq_id <- ((groupid - 1)*gene_num_in_array + 1):sub_seq_num\n",
    "}\n",
    "\n",
    "## exclude large coding masks\n",
    "if(arrayid==57)\n",
    "{\n",
    "  sub_seq_id <- setdiff(sub_seq_id,840)\n",
    "}\n",
    "\n",
    "if(arrayid==112)\n",
    "{\n",
    "  sub_seq_id <- setdiff(sub_seq_id,c(543,544))\n",
    "}\n",
    "\n",
    "if(arrayid==113)\n",
    "{\n",
    "  sub_seq_id <- setdiff(sub_seq_id,c(575,576,577,578,579,580,582))\n",
    "}\n",
    "\n",
    "genes <- genes_info\n",
    "\n",
    "results_coding <- c()\n",
    "\n",
    "UKB_SumStat.File.Path <- paste0(file.dir[1],file.prefix[1],\"_sumstat_\",arrayid,\"_470.Rdata\")\n",
    "AoU_SumStat.File.Path <- paste0(file.dir[2],file.prefix[2],\"_sumstat_\",arrayid,\".Rdata\")\n",
    "\n",
    "UKB_Cov.File.Path <- paste0(file.dir[1],file.prefix[1],\"_cov_\",arrayid,\"_470.Rdata\")\n",
    "AoU_Cov.File.Path <- paste0(file.dir[2],file.prefix[2],\"_cov_\",arrayid,\".Rdata\")\n",
    "\n",
    "sumstat.file.path <- c(UKB_SumStat.File.Path,AoU_SumStat.File.Path)\n",
    "cov.file.path <- c(UKB_Cov.File.Path,AoU_Cov.File.Path)\n",
    "coding_sumstat_list <- sapply(sumstat.file.path, function(x) mget(load(x)), simplify = TRUE)\n",
    "coding_cov_list <- sapply(cov.file.path, function(x) mget(load(x)), simplify = TRUE)\n",
    "\n",
    "for(kk in sub_seq_id)\n",
    "{\n",
    "  print(kk)\n",
    "  gene_name <- genes_info_chr[kk,1]\n",
    "  coding_sumstat_gene_list <- lapply(sumstat.file.path, function(x) {\n",
    "    coding_sumstat_list[[paste0(x,\".coding_sumstat\")]][[gene_name]]\n",
    "  })\n",
    "  coding_cov_gene_list <- lapply(cov.file.path, function(x) {\n",
    "    coding_cov_list[[paste0(x,\".coding_cov\")]][[gene_name]]\n",
    "  })\n",
    "  results <- coding_MetaSTAARlite(chr=chr,gene_name=gene_name,genes=genes,\n",
    "                                  sample.sizes=sample.sizes,coding_sumstat_gene_list=coding_sumstat_gene_list,\n",
    "                                  coding_cov_gene_list=coding_cov_gene_list,\n",
    "                                  cov_maf_cutoff=cov_maf_cutoff,\n",
    "                                  rare_maf_cutoff=0.01,rv_num_cutoff=2,\n",
    "                                  check_qc_label=TRUE,variant_type=variant_type,\n",
    "                                  Use_annotation_weights=Use_annotation_weights,Annotation_name=Annotation_name)\n",
    "  results_coding <- append(results_coding,results)\n",
    "}\n",
    "\n",
    "save(results_coding,file=paste0(output_path,\"/\",output_file_name,\"_\",arrayid,\".Rdata\"))\n",
    "})[3]\n",
    "\n",
    "save(time,file = paste0(output_path,\"/\",output_file_name,\"_time_\",arrayid,\".Rdata\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing Meta_Coding_UKB_AoU.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile Meta_Coding_UKB_AoU.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${arrayid} ${UKB_Sumstats} ${AoU_Sumstats} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "for(arrayid in c(380:381)){\n",
    "        tasks <- rbind(tasks, data.frame(\n",
    "            '--env arrayid'=arrayid,\n",
    "            '--input-recursive UKB_Sumstats'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/MetaSTAAR/UKB_Sumstats\",\n",
    "            '--input-recursive AoU_Sumstats'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/MetaSTAAR/AoU_Sumstats\",\n",
    "            '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/MetaSTAAR/Meta_Coding_UKB_AoU.R\",\n",
    "            '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/MetaSTAAR/UKB_AoU_Meta\",\n",
    "            check.names = FALSE\n",
    "        )) \n",
    "    }\n",
    "  \n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://Meta_Coding_UKB_AoU.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][ 22.0 KiB/ 22.0 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/22.0 KiB.                                     \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp Meta_Coding_UKB_AoU.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/MetaSTAAR/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: meta-codin--williamsjacr--241120-151534-74\n",
      "  job-name: meta-coding-ukb-aou\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/15867520956650507802\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/1469970834848913314\n",
      "Launched job-id: meta-codin--williamsjacr--241120-151534-74\n",
      "2 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'meta-codin--williamsjacr--241120-151534-74' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'meta-codin--williamsjacr--241120-151534-74' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script Meta_Coding_UKB_AoU.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status    Last Update\r\n",
      "---------------  ------  --------  --------------\r\n",
      "meta-coding-...       2  Success   11-20 15:20:41\r\n",
      "meta-coding-...       1  Success   11-20 15:20:26\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'meta-codin--williamsjacr--241120-151534-74' --users 'williamsjacr' --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "code_folding": [
     17
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Meta_Coding_UKB_AoU_Summary.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile Meta_Coding_UKB_AoU_Summary.R\n",
    "\n",
    "rm(list=ls())\n",
    "gc()\n",
    "\n",
    "Meta_SumStats <- commandArgs(TRUE)[1]\n",
    "print(Meta_SumStats)\n",
    "\n",
    "OUTPUT_PATH <- commandArgs(TRUE)[2]\n",
    "print(OUTPUT_PATH)\n",
    "\n",
    "## load required packages\n",
    "library(STAAR)\n",
    "library(STAARpipeline)\n",
    "library(STAARpipelineSummary)\n",
    "library(MetaSTAAR)\n",
    "\n",
    "Gene_Centric_Coding_Results_Summary_meta <- function(gene_centric_coding_jobs_num,input_path,output_path,gene_centric_results_name,\n",
    "                                                     alpha=2.5E-06,manhattan_plot=FALSE,QQ_plot=FALSE){\n",
    "\n",
    "  #######################################################\n",
    "  #     summarize unconditional analysis results\n",
    "  #######################################################\n",
    "\n",
    "  results_coding_genome <- c()\n",
    "\n",
    "  for(kk in 1:gene_centric_coding_jobs_num)\n",
    "  {\n",
    "    print(kk)\n",
    "    results_coding <- get(load(paste0(input_path,gene_centric_results_name,\"_\",kk,\".Rdata\")))\n",
    "\n",
    "    results_coding_genome <- c(results_coding_genome,results_coding)\n",
    "  }\n",
    "\n",
    "  results_plof_genome <- c()\n",
    "  results_plof_ds_genome <- c()\n",
    "  results_missense_genome <- c()\n",
    "  results_disruptive_missense_genome <- c()\n",
    "  results_synonymous_genome <- c()\n",
    "\n",
    "  for(kk in 1:length(results_coding_genome))\n",
    "  {\n",
    "    results <- results_coding_genome[[kk]]\n",
    "\n",
    "    if(is.null(results)==FALSE)\n",
    "    {\n",
    "      ### plof\n",
    "      if(results[3]==\"plof\")\n",
    "      {\n",
    "        results_plof_genome <- rbind(results_plof_genome,results)\n",
    "      }\n",
    "      ### plof_ds\n",
    "      if(results[3]==\"plof_ds\")\n",
    "      {\n",
    "        results_plof_ds_genome <- rbind(results_plof_ds_genome,results)\n",
    "      }\n",
    "      ### missense\n",
    "      if(results[3]==\"missense\")\n",
    "      {\n",
    "        results_missense_genome <- rbind(results_missense_genome,results)\n",
    "      }\n",
    "      ### disruptive_missense\n",
    "      if(results[3]==\"disruptive_missense\")\n",
    "      {\n",
    "        results_disruptive_missense_genome <- rbind(results_disruptive_missense_genome,results)\n",
    "      }\n",
    "      ### synonymous\n",
    "      if(results[3]==\"synonymous\")\n",
    "      {\n",
    "        results_synonymous_genome <- rbind(results_synonymous_genome,results)\n",
    "      }\n",
    "    }\n",
    "\n",
    "    if(kk%%1000==0)\n",
    "    {\n",
    "      print(kk)\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ###### whole-genome results\n",
    "  # plof\n",
    "  save(results_plof_genome,file=paste0(output_path,\"plof.Rdata\"))\n",
    "  # plof + disruptive missense\n",
    "  save(results_plof_ds_genome,file=paste0(output_path,\"plof_ds.Rdata\"))\n",
    "  # missense\n",
    "  save(results_missense_genome,file=paste0(output_path,\"missense.Rdata\"))\n",
    "  # disruptive missense\n",
    "  save(results_disruptive_missense_genome,file=paste0(output_path,\"disruptive_missense.Rdata\"))\n",
    "  # synonymous\n",
    "  save(results_synonymous_genome,file=paste0(output_path,\"synonymous.Rdata\"))\n",
    "\n",
    "  ###### significant results\n",
    "  # plof\n",
    "  plof_sig <- results_plof_genome[results_plof_genome[,\"MetaSTAAR-O\"]<alpha,,drop=FALSE]\n",
    "  write.csv(plof_sig,file=paste0(output_path,\"plof_sig.csv\"))\n",
    "  # missense\n",
    "  missense_sig <- results_missense_genome[results_missense_genome[,\"MetaSTAAR-O\"]<alpha,,drop=FALSE]\n",
    "  write.csv(missense_sig,file=paste0(output_path,\"missense_sig.csv\"))\n",
    "  # synonymous\n",
    "  synonymous_sig <- results_synonymous_genome[results_synonymous_genome[,\"MetaSTAAR-O\"]<alpha,,drop=FALSE]\n",
    "  write.csv(synonymous_sig,file=paste0(output_path,\"synonymous_sig.csv\"))\n",
    "  # plof_ds\n",
    "  plof_ds_sig <- results_plof_ds_genome[results_plof_ds_genome[,\"MetaSTAAR-O\"]<alpha,,drop=FALSE]\n",
    "  write.csv(plof_ds_sig,file=paste0(output_path,\"plof_ds_sig.csv\"))\n",
    "  # disruptive_missense\n",
    "  disruptive_missense_sig <- results_disruptive_missense_genome[results_disruptive_missense_genome[,\"MetaSTAAR-O\"]<alpha,,drop=FALSE]\n",
    "  write.csv(disruptive_missense_sig,file=paste0(output_path,\"disruptive_missense_sig.csv\"))\n",
    "  # coding results\n",
    "  coding_sig <- rbind(plof_sig[,c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"SKAT-MS(1,25)\",\"Burden-MS(1,1)\",\"ACAT-V-MS(1,25)\",\"MetaSTAAR-O\")],\n",
    "                      missense_sig[,c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"SKAT-MS(1,25)\",\"Burden-MS(1,1)\",\"ACAT-V-MS(1,25)\",\"MetaSTAAR-O\")])\n",
    "  coding_sig <- rbind(coding_sig,synonymous_sig[,c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"SKAT-MS(1,25)\",\"Burden-MS(1,1)\",\"ACAT-V-MS(1,25)\",\"MetaSTAAR-O\")])\n",
    "  coding_sig <- rbind(coding_sig,plof_ds_sig[,c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"SKAT-MS(1,25)\",\"Burden-MS(1,1)\",\"ACAT-V-MS(1,25)\",\"MetaSTAAR-O\")])\n",
    "  coding_sig <- rbind(coding_sig,disruptive_missense_sig[,c(\"Gene name\",\"Chr\",\"Category\",\"#SNV\",\"SKAT-MS(1,25)\",\"Burden-MS(1,1)\",\"ACAT-V-MS(1,25)\",\"MetaSTAAR-O\")])\n",
    "  write.csv(coding_sig,file=paste0(output_path,\"coding_sig.csv\"))\n",
    "\n",
    "  ## manhattan plot\n",
    "  if(manhattan_plot)\n",
    "  {\n",
    "    ### plof\n",
    "    results_MetaSTAAR <- results_plof_genome[,c(1,2,dim(results_plof_genome)[2])]\n",
    "\n",
    "    results_m <- c()\n",
    "    for(i in 1:dim(results_MetaSTAAR)[2])\n",
    "    {\n",
    "      results_m <- cbind(results_m,unlist(results_MetaSTAAR[,i]))\n",
    "    }\n",
    "\n",
    "    colnames(results_m) <- colnames(results_MetaSTAAR)\n",
    "    results_m <- data.frame(results_m,stringsAsFactors = FALSE)\n",
    "    results_m[,2] <- as.numeric(results_m[,2])\n",
    "    results_m[,3] <- as.numeric(results_m[,3])\n",
    "\n",
    "    genes_info_manhattan <- dplyr::left_join(genes_info,results_m,by=c(\"chromosome_name\"=\"Chr\",\"hgnc_symbol\"=\"Gene.name\"))\n",
    "\n",
    "    genes_info_manhattan[is.na(genes_info_manhattan)] <- 1\n",
    "    colnames(genes_info_manhattan)[dim(genes_info_manhattan)[2]] <- \"plof\"\n",
    "\n",
    "    ### plof_ds\n",
    "    results_MetaSTAAR <- results_plof_ds_genome[,c(1,2,dim(results_plof_ds_genome)[2])]\n",
    "\n",
    "    results_m <- c()\n",
    "    for(i in 1:dim(results_MetaSTAAR)[2])\n",
    "    {\n",
    "      results_m <- cbind(results_m,unlist(results_MetaSTAAR[,i]))\n",
    "    }\n",
    "\n",
    "    colnames(results_m) <- colnames(results_MetaSTAAR)\n",
    "    results_m <- data.frame(results_m,stringsAsFactors = FALSE)\n",
    "    results_m[,2] <- as.numeric(results_m[,2])\n",
    "    results_m[,3] <- as.numeric(results_m[,3])\n",
    "\n",
    "    genes_info_manhattan <- dplyr::left_join(genes_info_manhattan,results_m,by=c(\"chromosome_name\"=\"Chr\",\"hgnc_symbol\"=\"Gene.name\"))\n",
    "    genes_info_manhattan[is.na(genes_info_manhattan)] <- 1\n",
    "    colnames(genes_info_manhattan)[dim(genes_info_manhattan)[2]] <- \"plof_ds\"\n",
    "\n",
    "    ### missense\n",
    "    results_MetaSTAAR <- results_missense_genome[,c(1,2,dim(results_missense_genome)[2]-6)]\n",
    "\n",
    "    results_m <- c()\n",
    "    for(i in 1:dim(results_MetaSTAAR)[2])\n",
    "    {\n",
    "      results_m <- cbind(results_m,unlist(results_MetaSTAAR[,i]))\n",
    "    }\n",
    "\n",
    "    colnames(results_m) <- colnames(results_MetaSTAAR)\n",
    "    results_m <- data.frame(results_m,stringsAsFactors = FALSE)\n",
    "    results_m[,2] <- as.numeric(results_m[,2])\n",
    "    results_m[,3] <- as.numeric(results_m[,3])\n",
    "\n",
    "    genes_info_manhattan <- dplyr::left_join(genes_info_manhattan,results_m,by=c(\"chromosome_name\"=\"Chr\",\"hgnc_symbol\"=\"Gene.name\"))\n",
    "    genes_info_manhattan[is.na(genes_info_manhattan)] <- 1\n",
    "    colnames(genes_info_manhattan)[dim(genes_info_manhattan)[2]] <- \"missense\"\n",
    "\n",
    "    ### disruptive_missense\n",
    "    results_MetaSTAAR <- results_disruptive_missense_genome[,c(1,2,dim(results_disruptive_missense_genome)[2])]\n",
    "\n",
    "    results_m <- c()\n",
    "    for(i in 1:dim(results_MetaSTAAR)[2])\n",
    "    {\n",
    "      results_m <- cbind(results_m,unlist(results_MetaSTAAR[,i]))\n",
    "    }\n",
    "\n",
    "    colnames(results_m) <- colnames(results_MetaSTAAR)\n",
    "    results_m <- data.frame(results_m,stringsAsFactors = FALSE)\n",
    "    results_m[,2] <- as.numeric(results_m[,2])\n",
    "    results_m[,3] <- as.numeric(results_m[,3])\n",
    "\n",
    "    genes_info_manhattan <- dplyr::left_join(genes_info_manhattan,results_m,by=c(\"chromosome_name\"=\"Chr\",\"hgnc_symbol\"=\"Gene.name\"))\n",
    "    genes_info_manhattan[is.na(genes_info_manhattan)] <- 1\n",
    "    colnames(genes_info_manhattan)[dim(genes_info_manhattan)[2]] <- \"disruptive_missense\"\n",
    "\n",
    "    ### synonymous\n",
    "    results_MetaSTAAR <- results_synonymous_genome[,c(1,2,dim(results_synonymous_genome)[2])]\n",
    "\n",
    "    results_m <- c()\n",
    "    for(i in 1:dim(results_MetaSTAAR)[2])\n",
    "    {\n",
    "      results_m <- cbind(results_m,unlist(results_MetaSTAAR[,i]))\n",
    "    }\n",
    "\n",
    "    colnames(results_m) <- colnames(results_MetaSTAAR)\n",
    "    results_m <- data.frame(results_m,stringsAsFactors = FALSE)\n",
    "    results_m[,2] <- as.numeric(results_m[,2])\n",
    "    results_m[,3] <- as.numeric(results_m[,3])\n",
    "\n",
    "    genes_info_manhattan <- dplyr::left_join(genes_info_manhattan,results_m,by=c(\"chromosome_name\"=\"Chr\",\"hgnc_symbol\"=\"Gene.name\"))\n",
    "    genes_info_manhattan[is.na(genes_info_manhattan)] <- 1\n",
    "    colnames(genes_info_manhattan)[dim(genes_info_manhattan)[2]] <- \"synonymous\"\n",
    "\n",
    "    ## ylim\n",
    "    coding_minp <- min(genes_info_manhattan[,(dim(genes_info_manhattan)[2]-4):dim(genes_info_manhattan)[2]])\n",
    "    min_y <- ceiling(-log10(coding_minp)) + 1\n",
    "\n",
    "    pch <- c(0,1,2,3,4)\n",
    "    figure1 <- manhattan_plot(genes_info_manhattan[,2], (genes_info_manhattan[,3]+genes_info_manhattan[,4])/2, genes_info_manhattan$plof,sig.level=alpha,pch=pch[1],col = c(\"blue4\", \"orange3\"),ylim=c(0,min_y),\n",
    "                              auto.key=T,key=list(space=\"top\", columns=5, title=\"Functional Category\", cex.title=1, points=TRUE,pch=pch,text=list(c(\"pLoF\",\"pLoF+D\",\"Missense\",\"Disruptive Missense\",\"Synonymous\"))))\n",
    "\n",
    "    figure2 <- manhattan_plot(genes_info_manhattan[,2], (genes_info_manhattan[,3]+genes_info_manhattan[,4])/2, genes_info_manhattan$plof_ds,sig.level=alpha,pch=pch[2],col = c(\"blue4\", \"orange3\"),ylim=c(0,min_y),\n",
    "                              auto.key=T,key=list(space=\"top\", columns=5, title=\"Functional Category\", cex.title=1, points=TRUE,pch=pch,text=list(c(\"pLoF\",\"pLoF+D\",\"Missense\",\"Disruptive Missense\",\"Synonymous\"))))\n",
    "\n",
    "    figure3 <- manhattan_plot(genes_info_manhattan[,2], (genes_info_manhattan[,3]+genes_info_manhattan[,4])/2, genes_info_manhattan$missense,sig.level=alpha,pch=pch[3],col = c(\"blue4\", \"orange3\"),ylim=c(0,min_y),\n",
    "                              auto.key=T,key=list(space=\"top\", columns=5, title=\"Functional Category\", cex.title=1, points=TRUE,pch=pch,text=list(c(\"pLoF\",\"pLoF+D\",\"Missense\",\"Disruptive Missense\",\"Synonymous\"))))\n",
    "\n",
    "    figure4 <- manhattan_plot(genes_info_manhattan[,2], (genes_info_manhattan[,3]+genes_info_manhattan[,4])/2, genes_info_manhattan$disruptive_missense,sig.level=alpha,pch=pch[4],col = c(\"blue4\", \"orange3\"),ylim=c(0,min_y),\n",
    "                              auto.key=T,key=list(space=\"top\", columns=5, title=\"Functional Category\", cex.title=1, points=TRUE,pch=pch,text=list(c(\"pLoF\",\"pLoF+D\",\"Missense\",\"Disruptive Missense\",\"Synonymous\"))))\n",
    "\n",
    "    figure5 <- manhattan_plot(genes_info_manhattan[,2], (genes_info_manhattan[,3]+genes_info_manhattan[,4])/2, genes_info_manhattan$synonymous,sig.level=alpha,pch=pch[5],col = c(\"blue4\", \"orange3\"),ylim=c(0,min_y),\n",
    "                              auto.key=T,key=list(space=\"top\", columns=5, title=\"Functional Category\", cex.title=1, points=TRUE,pch=pch,text=list(c(\"pLoF\",\"pLoF+D\",\"Missense\",\"Disruptive Missense\",\"Synonymous\"))))\n",
    "\n",
    "    print(\"Manhattan plot\")\n",
    "\n",
    "    png(paste0(output_path,\"gene_centric_coding_manhattan.png\"), width = 9, height = 6, units = 'in', res = 600)\n",
    "\n",
    "    print(figure1)\n",
    "    print(figure2,newpage = FALSE)\n",
    "    print(figure3,newpage = FALSE)\n",
    "    print(figure4,newpage = FALSE)\n",
    "    print(figure5,newpage = FALSE)\n",
    "\n",
    "    dev.off()\n",
    "  }\n",
    "\n",
    "  ## Q-Q plot\n",
    "  if(QQ_plot)\n",
    "  {\n",
    "    if(!manhattan_plot)\n",
    "    {\n",
    "      ### plof\n",
    "      results_MetaSTAAR <- results_plof_genome[,c(1,2,dim(results_plof_genome)[2])]\n",
    "\n",
    "      results_m <- c()\n",
    "      for(i in 1:dim(results_MetaSTAAR)[2])\n",
    "      {\n",
    "        results_m <- cbind(results_m,unlist(results_MetaSTAAR[,i]))\n",
    "      }\n",
    "\n",
    "      colnames(results_m) <- colnames(results_MetaSTAAR)\n",
    "      results_m <- data.frame(results_m,stringsAsFactors = FALSE)\n",
    "      results_m[,2] <- as.numeric(results_m[,2])\n",
    "      results_m[,3] <- as.numeric(results_m[,3])\n",
    "\n",
    "      genes_info_manhattan <- dplyr::left_join(genes_info,results_m,by=c(\"chromosome_name\"=\"Chr\",\"hgnc_symbol\"=\"Gene.name\"))\n",
    "\n",
    "      genes_info_manhattan[is.na(genes_info_manhattan)] <- 1\n",
    "      colnames(genes_info_manhattan)[dim(genes_info_manhattan)[2]] <- \"plof\"\n",
    "\n",
    "      ### plof_ds\n",
    "      results_MetaSTAAR <- results_plof_ds_genome[,c(1,2,dim(results_plof_ds_genome)[2])]\n",
    "\n",
    "      results_m <- c()\n",
    "      for(i in 1:dim(results_MetaSTAAR)[2])\n",
    "      {\n",
    "        results_m <- cbind(results_m,unlist(results_MetaSTAAR[,i]))\n",
    "      }\n",
    "\n",
    "      colnames(results_m) <- colnames(results_MetaSTAAR)\n",
    "      results_m <- data.frame(results_m,stringsAsFactors = FALSE)\n",
    "      results_m[,2] <- as.numeric(results_m[,2])\n",
    "      results_m[,3] <- as.numeric(results_m[,3])\n",
    "\n",
    "      genes_info_manhattan <- dplyr::left_join(genes_info_manhattan,results_m,by=c(\"chromosome_name\"=\"Chr\",\"hgnc_symbol\"=\"Gene.name\"))\n",
    "      genes_info_manhattan[is.na(genes_info_manhattan)] <- 1\n",
    "      colnames(genes_info_manhattan)[dim(genes_info_manhattan)[2]] <- \"plof_ds\"\n",
    "\n",
    "      ### missense\n",
    "      results_MetaSTAAR <- results_missense_genome[,c(1,2,dim(results_missense_genome)[2]-6)]\n",
    "\n",
    "      results_m <- c()\n",
    "      for(i in 1:dim(results_MetaSTAAR)[2])\n",
    "      {\n",
    "        results_m <- cbind(results_m,unlist(results_MetaSTAAR[,i]))\n",
    "      }\n",
    "\n",
    "      colnames(results_m) <- colnames(results_MetaSTAAR)\n",
    "      results_m <- data.frame(results_m,stringsAsFactors = FALSE)\n",
    "      results_m[,2] <- as.numeric(results_m[,2])\n",
    "      results_m[,3] <- as.numeric(results_m[,3])\n",
    "\n",
    "      genes_info_manhattan <- dplyr::left_join(genes_info_manhattan,results_m,by=c(\"chromosome_name\"=\"Chr\",\"hgnc_symbol\"=\"Gene.name\"))\n",
    "      genes_info_manhattan[is.na(genes_info_manhattan)] <- 1\n",
    "      colnames(genes_info_manhattan)[dim(genes_info_manhattan)[2]] <- \"missense\"\n",
    "\n",
    "      ### disruptive_missense\n",
    "      results_MetaSTAAR <- results_disruptive_missense_genome[,c(1,2,dim(results_disruptive_missense_genome)[2])]\n",
    "\n",
    "      results_m <- c()\n",
    "      for(i in 1:dim(results_MetaSTAAR)[2])\n",
    "      {\n",
    "        results_m <- cbind(results_m,unlist(results_MetaSTAAR[,i]))\n",
    "      }\n",
    "\n",
    "      colnames(results_m) <- colnames(results_MetaSTAAR)\n",
    "      results_m <- data.frame(results_m,stringsAsFactors = FALSE)\n",
    "      results_m[,2] <- as.numeric(results_m[,2])\n",
    "      results_m[,3] <- as.numeric(results_m[,3])\n",
    "\n",
    "      genes_info_manhattan <- dplyr::left_join(genes_info_manhattan,results_m,by=c(\"chromosome_name\"=\"Chr\",\"hgnc_symbol\"=\"Gene.name\"))\n",
    "      genes_info_manhattan[is.na(genes_info_manhattan)] <- 1\n",
    "      colnames(genes_info_manhattan)[dim(genes_info_manhattan)[2]] <- \"disruptive_missense\"\n",
    "\n",
    "      ### synonymous\n",
    "      results_MetaSTAAR <- results_synonymous_genome[,c(1,2,dim(results_synonymous_genome)[2])]\n",
    "\n",
    "      results_m <- c()\n",
    "      for(i in 1:dim(results_MetaSTAAR)[2])\n",
    "      {\n",
    "        results_m <- cbind(results_m,unlist(results_MetaSTAAR[,i]))\n",
    "      }\n",
    "\n",
    "      colnames(results_m) <- colnames(results_MetaSTAAR)\n",
    "      results_m <- data.frame(results_m,stringsAsFactors = FALSE)\n",
    "      results_m[,2] <- as.numeric(results_m[,2])\n",
    "      results_m[,3] <- as.numeric(results_m[,3])\n",
    "\n",
    "      genes_info_manhattan <- dplyr::left_join(genes_info_manhattan,results_m,by=c(\"chromosome_name\"=\"Chr\",\"hgnc_symbol\"=\"Gene.name\"))\n",
    "      genes_info_manhattan[is.na(genes_info_manhattan)] <- 1\n",
    "      colnames(genes_info_manhattan)[dim(genes_info_manhattan)[2]] <- \"synonymous\"\n",
    "\n",
    "      ## ylim\n",
    "      coding_minp <- min(genes_info_manhattan[,(dim(genes_info_manhattan)[2]-4):dim(genes_info_manhattan)[2]])\n",
    "      min_y <- ceiling(-log10(coding_minp)) + 1\n",
    "    }\n",
    "\n",
    "    print(\"Q-Q plot\")\n",
    "    cex_point <- 1\n",
    "\n",
    "    png(paste0(output_path,\"gene_centric_coding_qqplot.png\"), width = 9, height = 9, units = 'in', res = 600)\n",
    "\n",
    "    ### plof\n",
    "    ## remove unconverged p-values\n",
    "    observed <- sort(genes_info_manhattan$plof[genes_info_manhattan$plof < 1])\n",
    "    lobs <- -(log10(observed))\n",
    "\n",
    "    expected <- c(1:length(observed))\n",
    "    lexp <- -(log10(expected / (length(expected)+1)))\n",
    "\n",
    "    # par(mar=c(5,6,4,4))\n",
    "    plot(lexp,lobs,pch=0, cex=cex_point, xlim = c(0, 5), ylim = c(0, min_y),\n",
    "         xlab = expression(Expected ~ ~-log[10](italic(p))), ylab = expression(Observed ~ ~-log[10](italic(p))),\n",
    "         font.lab=2,cex.lab=1,cex.axis=1,font.axis=2)\n",
    "\n",
    "    abline(0, 1, col=\"red\",lwd=1)\n",
    "\n",
    "    ### plof_ds\n",
    "    ## remove unconverged p-values\n",
    "    observed <- sort(genes_info_manhattan$plof_ds[genes_info_manhattan$plof_ds < 1])\n",
    "\n",
    "    lobs <- -(log10(observed))\n",
    "\n",
    "    expected <- c(1:length(observed))\n",
    "    lexp <- -(log10(expected / (length(expected)+1)))\n",
    "\n",
    "    par(new=T)\n",
    "    # par(mar=c(5,6,4,4))\n",
    "    plot(lexp,lobs,pch=1, cex=cex_point, xlim = c(0, 5), ylim = c(0, min_y),\n",
    "         xlab = expression(Expected ~ ~-log[10](italic(p))), ylab = expression(Observed ~ ~-log[10](italic(p))),\n",
    "         font.lab=2,cex.lab=1,cex.axis=1,font.axis=2)\n",
    "\n",
    "    abline(0, 1, col=\"red\",lwd=1)\n",
    "\n",
    "    ### missense\n",
    "    ## remove unconverged p-values\n",
    "    observed <- sort(genes_info_manhattan$missense[genes_info_manhattan$missense < 1])\n",
    "\n",
    "    lobs <- -(log10(observed))\n",
    "\n",
    "    expected <- c(1:length(observed))\n",
    "    lexp <- -(log10(expected / (length(expected)+1)))\n",
    "\n",
    "    par(new=T)\n",
    "    # par(mar=c(5,6,4,4))\n",
    "    plot(lexp,lobs,pch=2, cex=cex_point, xlim = c(0, 5), ylim = c(0, min_y),\n",
    "         xlab = expression(Expected ~ ~-log[10](italic(p))), ylab = expression(Observed ~ ~-log[10](italic(p))),\n",
    "         font.lab=2,cex.lab=1,cex.axis=1,font.axis=2)\n",
    "\n",
    "    abline(0, 1, col=\"red\",lwd=1)\n",
    "\n",
    "    ### disruptive_missense\n",
    "    ## remove unconverged p-values\n",
    "    observed <- sort(genes_info_manhattan$disruptive_missense[genes_info_manhattan$disruptive_missense < 1])\n",
    "\n",
    "\n",
    "    lobs <- -(log10(observed))\n",
    "\n",
    "    expected <- c(1:length(observed))\n",
    "    lexp <- -(log10(expected / (length(expected)+1)))\n",
    "\n",
    "    par(new=T)\n",
    "    # par(mar=c(5,6,4,4))\n",
    "    plot(lexp,lobs,pch=3, cex=cex_point, xlim = c(0, 5), ylim = c(0, min_y),\n",
    "         xlab = expression(Expected ~ ~-log[10](italic(p))), ylab = expression(Observed ~ ~-log[10](italic(p))),\n",
    "         font.lab=2,cex.lab=1,cex.axis=1,font.axis=2)\n",
    "\n",
    "    abline(0, 1, col=\"red\",lwd=1)\n",
    "\n",
    "    ### synonymous\n",
    "    ## remove unconverged p-values\n",
    "    observed <- sort(genes_info_manhattan$synonymous[genes_info_manhattan$synonymous < 1])\n",
    "\n",
    "    lobs <- -(log10(observed))\n",
    "\n",
    "    expected <- c(1:length(observed))\n",
    "    lexp <- -(log10(expected / (length(expected)+1)))\n",
    "\n",
    "    par(new=T)\n",
    "    # par(mar=c(5,6,4,4))\n",
    "    plot(lexp,lobs,pch=4, cex=cex_point, xlim = c(0, 5), ylim = c(0, min_y),\n",
    "         xlab = expression(Expected ~ ~-log[10](italic(p))), ylab = expression(Observed ~ ~-log[10](italic(p))),\n",
    "         font.lab=2,cex.lab=1,cex.axis=1,font.axis=2)\n",
    "\n",
    "    abline(0, 1, col=\"red\",lwd=1)\n",
    "\n",
    "    legend(\"topleft\",legend=c(\"pLoF\",\"pLoF+D\",\"Missense\",\"Disruptive Missense\",\"Synonymous\"),ncol=1,bty=\"o\",box.lwd=1,pch=0:4,cex=1,text.font=2)\n",
    "\n",
    "    dev.off()\n",
    "  }\n",
    "\n",
    "}\n",
    "\n",
    "###########################################################\n",
    "#           User Input\n",
    "###########################################################\n",
    "## results path\n",
    "input_path <- Meta_SumStats\n",
    "output_path <- paste0(OUTPUT_PATH,\"/\")\n",
    "## number of jobs\n",
    "gene_centric_coding_jobs_num <- 381\n",
    "## results name\n",
    "gene_centric_results_name <- \"/UKB_AoU_Coding\"\n",
    "\n",
    "## alpha level\n",
    "alpha <- 5E-07\n",
    "\n",
    "###########################################################\n",
    "#           Main Function\n",
    "###########################################################\n",
    "Gene_Centric_Coding_Results_Summary_meta(gene_centric_coding_jobs_num=gene_centric_coding_jobs_num,\n",
    "                                         input_path=input_path,output_path=output_path,\n",
    "                                         gene_centric_results_name=gene_centric_results_name,\n",
    "                                         alpha=alpha,manhattan_plot=TRUE,QQ_plot=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Meta_Coding_UKB_AoU_Summary.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile Meta_Coding_UKB_AoU_Summary.sh\n",
    "#!/bin/bash\n",
    "\n",
    "set -o errexit\n",
    "set -o nounset\n",
    "\n",
    "Rscript ${R_Script} ${Meta_SumStats} ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting score_task.R\n"
     ]
    }
   ],
   "source": [
    "%%writefile score_task.R\n",
    "\n",
    "tasks <- data.frame(check.names = FALSE)\n",
    "\n",
    "\n",
    "tasks <- rbind(tasks, data.frame(\n",
    "            '--input-recursive Meta_SumStats'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/MetaSTAAR/UKB_AoU_Meta\",\n",
    "            '--input R_Script'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/MetaSTAAR/Meta_Coding_UKB_AoU_Summary.R\",\n",
    "            '--output-recursive OUTPUT_PATH'=\"gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/MetaSTAAR/Meta_Results\",\n",
    "            check.names = FALSE\n",
    "        )) \n",
    "    \n",
    "  \n",
    "write.table(tasks, \n",
    "            file=\"score_task.txt\", \n",
    "            row.names=F, col.names=T, \n",
    "            sep='\\t', quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "!Rscript score_task.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://Meta_Coding_UKB_AoU_Summary.R [Content-Type=application/octet-stream]...\n",
      "/ [1/1 files][ 18.6 KiB/ 18.6 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/18.6 KiB.                                     \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp Meta_Coding_UKB_AoU_Summary.R gs://fc-secure-797107a7-4402-4122-941c-9a486e0d633e/MetaSTAAR/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Job properties:\n",
      "  job-id: meta-codin--williamsjacr--241120-152842-90\n",
      "  job-name: meta-coding-ukb-aou-summary\n",
      "  user-id: williamsjacr\n",
      "Provider internal-id (operation): projects/52933917155/locations/us-central1/operations/3908227709257602441\n",
      "Launched job-id: meta-codin--williamsjacr--241120-152842-90\n",
      "1 task(s)\n",
      "To check the status, run:\n",
      "  dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'meta-codin--williamsjacr--241120-152842-90' --users 'williamsjacr' --status '*'\n",
      "To cancel the job, run:\n",
      "  ddel --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'meta-codin--williamsjacr--241120-152842-90' --users 'williamsjacr'\n"
     ]
    }
   ],
   "source": [
    "%%bash --out score_batch\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "aou_dsub \\\n",
    "  --image willja16/r_with_plink \\\n",
    "  --disk-size 100 \\\n",
    "  --boot-disk-size 25 \\\n",
    "  --min-ram 16 \\\n",
    "  --timeout \"96h\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --script Meta_Coding_UKB_AoU_Summary.sh \\\n",
    "  --tasks score_task.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job Name           Task  Status    Last Update\r\n",
      "---------------  ------  --------  --------------\r\n",
      "meta-coding-...       1  Success   11-20 16:28:29\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!dstat --provider google-cls-v2 --project terra-vpc-sc-804f445b --location us-central1 --jobs 'meta-codin--williamsjacr--241120-152842-90' --users 'williamsjacr' --status '*'"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
